{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistical analysis to get the significantly associated tokens for attribution score groupings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read mapping information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>visit_occurrence_id</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>SEX</th>\n",
       "      <th>Age_Bin</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>678487</td>\n",
       "      <td>3312694</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>3312694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2094045</td>\n",
       "      <td>9199249</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(20, 30]</td>\n",
       "      <td>9199249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2107813</td>\n",
       "      <td>9410987</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(50, 60]</td>\n",
       "      <td>9410987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2083212</td>\n",
       "      <td>602890</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2083212</td>\n",
       "      <td>9039437</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7194</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16588204</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16588204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7195</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16709341</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16709341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7196</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16651755</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16651755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7197</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16872774</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16872774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16869916</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16869916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7199 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      person_id  visit_occurrence_id       TYPE COUNTRY_OF_BIRTH     SEX  \\\n",
       "0        678487              3312694  Emergency        Australia    Male   \n",
       "1       2094045              9199249  Emergency        Australia    Male   \n",
       "2       2107813              9410987  Emergency        Australia    Male   \n",
       "3       2083212               602890  Inpatient        Australia  Female   \n",
       "4       2083212              9039437  Emergency        Australia  Female   \n",
       "...         ...                  ...        ...              ...     ...   \n",
       "7194    2603535             16588204  Inpatient        Australia    Male   \n",
       "7195    2603535             16709341  Inpatient        Australia    Male   \n",
       "7196    2603535             16651755  Inpatient        Australia    Male   \n",
       "7197    2098616             16872774  Emergency        Australia  Female   \n",
       "7198    2098616             16869916  Emergency        Australia  Female   \n",
       "\n",
       "       Age_Bin  JOURNEY_ID  \n",
       "0     (60, 70]     3312694  \n",
       "1     (20, 30]     9199249  \n",
       "2     (50, 60]     9410987  \n",
       "3     (40, 50]     9039437  \n",
       "4     (40, 50]     9039437  \n",
       "...        ...         ...  \n",
       "7194  (60, 70]    16588204  \n",
       "7195  (60, 70]    16709341  \n",
       "7196  (60, 70]    16651755  \n",
       "7197  (80, 90]    16872774  \n",
       "7198  (80, 90]    16869916  \n",
       "\n",
       "[7199 rows x 7 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "demoTrainDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_train.csv', sep='\\t')\n",
    "demoTestDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_test.csv', sep='\\t')\n",
    "demoValidateDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_validate.csv', sep='\\t')\n",
    "demoDf = pd.concat([demoTrainDf, demoTestDf, demoValidateDf], ignore_index=True)\n",
    "demoDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tube_code</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>db_ID</th>\n",
       "      <th>pt_age</th>\n",
       "      <th>pt_gender</th>\n",
       "      <th>EPISODE_ID</th>\n",
       "      <th>hospital_admission</th>\n",
       "      <th>hospital_discharge</th>\n",
       "      <th>hospital_in_last_year</th>\n",
       "      <th>genome_species</th>\n",
       "      <th>species_reported</th>\n",
       "      <th>contig_number</th>\n",
       "      <th>length</th>\n",
       "      <th>tube_code_duplicate_or_old</th>\n",
       "      <th>date_of_collection</th>\n",
       "      <th>ID_number</th>\n",
       "      <th>location_additional_02</th>\n",
       "      <th>collected_from_original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH19J072</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>92</td>\n",
       "      <td>5131021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-0997</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH19J074</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus sp. D19</td>\n",
       "      <td>Streptococcus mitis group</td>\n",
       "      <td>84</td>\n",
       "      <td>2022991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-1174</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH19B003</td>\n",
       "      <td>526238.0</td>\n",
       "      <td>567AE</td>\n",
       "      <td>85</td>\n",
       "      <td>Male</td>\n",
       "      <td>640261.0</td>\n",
       "      <td>2019-02-04</td>\n",
       "      <td>2019-02-07</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>93</td>\n",
       "      <td>4942915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-02-03</td>\n",
       "      <td>19-034-1596</td>\n",
       "      <td>A-4WB;GMC</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH20A024</td>\n",
       "      <td>788941.0</td>\n",
       "      <td>3YBDS</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>13194824.0</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>110</td>\n",
       "      <td>5149157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>20-016-2112</td>\n",
       "      <td>S-ED;Emergency</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH20L041</td>\n",
       "      <td>1023735.0</td>\n",
       "      <td>2VJ78</td>\n",
       "      <td>42</td>\n",
       "      <td>Male</td>\n",
       "      <td>14210147.0</td>\n",
       "      <td>2020-12-18</td>\n",
       "      <td>2021-01-05</td>\n",
       "      <td>yes</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>46</td>\n",
       "      <td>2735216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-16</td>\n",
       "      <td>20-351-2856</td>\n",
       "      <td>Emergency Dept;nan</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3112</th>\n",
       "      <td>ALF22L076</td>\n",
       "      <td>639020.0</td>\n",
       "      <td>3S6HL</td>\n",
       "      <td>84</td>\n",
       "      <td>Male</td>\n",
       "      <td>17219899.0</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-20</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus vestibularis</td>\n",
       "      <td>Streptococcus salivarius group</td>\n",
       "      <td>107</td>\n",
       "      <td>1938627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>22-349-2150</td>\n",
       "      <td>Ward 4WB</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>ALF22L089</td>\n",
       "      <td>655404.0</td>\n",
       "      <td>6EWBM</td>\n",
       "      <td>67</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus columbae</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>152</td>\n",
       "      <td>3972219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-17</td>\n",
       "      <td>22-351-1465</td>\n",
       "      <td>Comm/Amb Clinic</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3114</th>\n",
       "      <td>ALF22L138</td>\n",
       "      <td>2572048.0</td>\n",
       "      <td>YM5PJ</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter cloacae complex</td>\n",
       "      <td>66</td>\n",
       "      <td>4876179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-27</td>\n",
       "      <td>22-361-0202</td>\n",
       "      <td>Ward 7 East</td>\n",
       "      <td>Hick White Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3115</th>\n",
       "      <td>ALF23A102</td>\n",
       "      <td>2054964.0</td>\n",
       "      <td>5DYAH</td>\n",
       "      <td>35</td>\n",
       "      <td>Female</td>\n",
       "      <td>17294287.0</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus constellatus</td>\n",
       "      <td>Streptococcus anginosus group</td>\n",
       "      <td>54</td>\n",
       "      <td>1869828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-11</td>\n",
       "      <td>23-011-3305</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>ALF23B102</td>\n",
       "      <td>2015952.0</td>\n",
       "      <td>57E2W</td>\n",
       "      <td>82</td>\n",
       "      <td>Female</td>\n",
       "      <td>17281239.0</td>\n",
       "      <td>2023-01-08</td>\n",
       "      <td>2023-02-16</td>\n",
       "      <td>no</td>\n",
       "      <td>Acinetobacter colistiniresistens</td>\n",
       "      <td>Acinetobacter species</td>\n",
       "      <td>233</td>\n",
       "      <td>3952992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-15</td>\n",
       "      <td>23-046-3310</td>\n",
       "      <td>ICU</td>\n",
       "      <td>PICC Line White</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3117 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      tube_code  PATIENT_ID  db_ID  pt_age pt_gender  EPISODE_ID  \\\n",
       "0      AH19J072    213972.0  3JNQS      64      Male  12931474.0   \n",
       "1      AH19J074    213972.0  3JNQS      64      Male  12931474.0   \n",
       "2      AH19B003    526238.0  567AE      85      Male    640261.0   \n",
       "3      AH20A024    788941.0  3YBDS      58    Female  13194824.0   \n",
       "4      AH20L041   1023735.0  2VJ78      42      Male  14210147.0   \n",
       "...         ...         ...    ...     ...       ...         ...   \n",
       "3112  ALF22L076    639020.0  3S6HL      84      Male  17219899.0   \n",
       "3113  ALF22L089    655404.0  6EWBM      67    Female         NaN   \n",
       "3114  ALF22L138   2572048.0  YM5PJ      61      Male         NaN   \n",
       "3115  ALF23A102   2054964.0  5DYAH      35    Female  17294287.0   \n",
       "3116  ALF23B102   2015952.0  57E2W      82    Female  17281239.0   \n",
       "\n",
       "     hospital_admission hospital_discharge hospital_in_last_year  \\\n",
       "0            2019-10-17         2019-11-29                   yes   \n",
       "1            2019-10-17         2019-11-29                   yes   \n",
       "2            2019-02-04         2019-02-07                    no   \n",
       "3            2020-01-16         2020-01-19                    no   \n",
       "4            2020-12-18         2021-01-05                   yes   \n",
       "...                 ...                ...                   ...   \n",
       "3112         2022-12-15         2022-12-20                   yes   \n",
       "3113                NaN                NaN                   NaN   \n",
       "3114                NaN                NaN                   NaN   \n",
       "3115         2023-01-12         2023-02-01                   yes   \n",
       "3116         2023-01-08         2023-02-16                    no   \n",
       "\n",
       "                        genome_species                species_reported  \\\n",
       "0                     Escherichia coli                Escherichia coli   \n",
       "1                Streptococcus sp. D19       Streptococcus mitis group   \n",
       "2                     Escherichia coli                Escherichia coli   \n",
       "3                     Escherichia coli                Escherichia coli   \n",
       "4                Staphylococcus aureus           Staphylococcus aureus   \n",
       "...                                ...                             ...   \n",
       "3112        Streptococcus vestibularis  Streptococcus salivarius group   \n",
       "3113                  Proteus columbae                Proteus vulgaris   \n",
       "3114         Enterobacter roggenkampii    Enterobacter cloacae complex   \n",
       "3115        Streptococcus constellatus   Streptococcus anginosus group   \n",
       "3116  Acinetobacter colistiniresistens           Acinetobacter species   \n",
       "\n",
       "      contig_number   length tube_code_duplicate_or_old date_of_collection  \\\n",
       "0                92  5131021                        NaN         2019-10-30   \n",
       "1                84  2022991                        NaN         2019-10-30   \n",
       "2                93  4942915                        NaN         2019-02-03   \n",
       "3               110  5149157                        NaN         2020-01-16   \n",
       "4                46  2735216                        NaN         2020-12-16   \n",
       "...             ...      ...                        ...                ...   \n",
       "3112            107  1938627                        NaN         2022-12-15   \n",
       "3113            152  3972219                        NaN         2022-12-17   \n",
       "3114             66  4876179                        NaN         2022-12-27   \n",
       "3115             54  1869828                        NaN         2023-01-11   \n",
       "3116            233  3952992                        NaN         2023-02-15   \n",
       "\n",
       "        ID_number location_additional_02 collected_from_original  \n",
       "0     19-303-0997             A-7EA;HAEM                   Blood  \n",
       "1     19-303-1174             A-7EA;HAEM        Blood Peripheral  \n",
       "2     19-034-1596              A-4WB;GMC                   Blood  \n",
       "3     20-016-2112         S-ED;Emergency            Blood Venous  \n",
       "4     20-351-2856     Emergency Dept;nan                   Blood  \n",
       "...           ...                    ...                     ...  \n",
       "3112  22-349-2150               Ward 4WB        Blood Peripheral  \n",
       "3113  22-351-1465        Comm/Amb Clinic                   Blood  \n",
       "3114  22-361-0202            Ward 7 East          Hick White Lum  \n",
       "3115  23-011-3305                    ICU            Blood Venous  \n",
       "3116  23-046-3310                    ICU         PICC Line White  \n",
       "\n",
       "[3117 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/patient_tube_id_mapping_full.tsv', sep='\\t')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "      <th>tube_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50056</td>\n",
       "      <td>13224842</td>\n",
       "      <td>AH20B011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62567</td>\n",
       "      <td>12645629</td>\n",
       "      <td>AH19G065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138108</td>\n",
       "      <td>14737781</td>\n",
       "      <td>AH21E052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>AH20J009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>KPN2214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>2084373</td>\n",
       "      <td>16158768</td>\n",
       "      <td>ALF22C004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>1058066</td>\n",
       "      <td>16759482</td>\n",
       "      <td>ALF22H089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>2552781</td>\n",
       "      <td>16966609</td>\n",
       "      <td>ALF22J054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2638484</td>\n",
       "      <td>16969999</td>\n",
       "      <td>ALF22K143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>2015952</td>\n",
       "      <td>17281239</td>\n",
       "      <td>ALF23B102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>517 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     person_id  JOURNEY_ID  tube_code\n",
       "0        50056    13224842   AH20B011\n",
       "1        62567    12645629   AH19G065\n",
       "2       138108    14737781   AH21E052\n",
       "3       211314    13955099   AH20J009\n",
       "4       211314    13955099    KPN2214\n",
       "..         ...         ...        ...\n",
       "512    2084373    16158768  ALF22C004\n",
       "513    1058066    16759482  ALF22H089\n",
       "514    2552781    16966609  ALF22J054\n",
       "515    2638484    16969999  ALF22K143\n",
       "516    2015952    17281239  ALF23B102\n",
       "\n",
       "[517 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mappedJourneyTubeIdsDf = mappingDf[['tube_code', 'PATIENT_ID', 'EPISODE_ID']].drop_duplicates().merge(\n",
    "    demoDf[['person_id', 'visit_occurrence_id', 'JOURNEY_ID']],\n",
    "    how='inner',\n",
    "    left_on=['PATIENT_ID', 'EPISODE_ID'],\n",
    "    right_on=['person_id', 'visit_occurrence_id']\n",
    ").drop(\n",
    "    columns=['PATIENT_ID', 'EPISODE_ID', 'visit_occurrence_id']\n",
    ")[['person_id', 'JOURNEY_ID', 'tube_code']].drop_duplicates()\n",
    "mappedJourneyTubeIdsDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read bed files containing overlapping tokens and attribution scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>tokens</th>\n",
       "      <th>score</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>atributes</th>\n",
       "      <th>tube_code</th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>193</td>\n",
       "      <td>197</td>\n",
       "      <td>ATAAC</td>\n",
       "      <td>-0.005218</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>204</td>\n",
       "      <td>210</td>\n",
       "      <td>ATTGATT</td>\n",
       "      <td>-0.004120</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>211</td>\n",
       "      <td>218</td>\n",
       "      <td>AATATTTT</td>\n",
       "      <td>0.003234</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>219</td>\n",
       "      <td>225</td>\n",
       "      <td>GTCATTG</td>\n",
       "      <td>-0.009522</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>226</td>\n",
       "      <td>233</td>\n",
       "      <td>AATTCTAC</td>\n",
       "      <td>-0.004358</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94483</th>\n",
       "      <td>121154</td>\n",
       "      <td>20</td>\n",
       "      <td>410</td>\n",
       "      <td>415</td>\n",
       "      <td>TAACTT</td>\n",
       "      <td>0.017504</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_13685</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>yopx</td>\n",
       "      <td>ID=KKIOAG_13685;Name=YopX domain-containing pr...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94484</th>\n",
       "      <td>121155</td>\n",
       "      <td>20</td>\n",
       "      <td>416</td>\n",
       "      <td>420</td>\n",
       "      <td>TTTTA</td>\n",
       "      <td>0.001789</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_13685</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>yopx</td>\n",
       "      <td>ID=KKIOAG_13685;Name=YopX domain-containing pr...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94485</th>\n",
       "      <td>121156</td>\n",
       "      <td>20</td>\n",
       "      <td>421</td>\n",
       "      <td>427</td>\n",
       "      <td>TCTTTAT</td>\n",
       "      <td>0.003276</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_13685</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>yopx</td>\n",
       "      <td>ID=KKIOAG_13685;Name=YopX domain-containing pr...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94486</th>\n",
       "      <td>121157</td>\n",
       "      <td>20</td>\n",
       "      <td>428</td>\n",
       "      <td>432</td>\n",
       "      <td>CCCAC</td>\n",
       "      <td>0.005503</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_13685</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>yopx</td>\n",
       "      <td>ID=KKIOAG_13685;Name=YopX domain-containing pr...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94487</th>\n",
       "      <td>121159</td>\n",
       "      <td>20</td>\n",
       "      <td>437</td>\n",
       "      <td>443</td>\n",
       "      <td>TAAATTT</td>\n",
       "      <td>0.009119</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_13685</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>yopx</td>\n",
       "      <td>ID=KKIOAG_13685;Name=YopX domain-containing pr...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73198 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index  contig_id  start_position  end_position    tokens     score  \\\n",
       "0           1          1             193           197     ATAAC -0.005218   \n",
       "1           4          1             204           210   ATTGATT -0.004120   \n",
       "2           5          1             211           218  AATATTTT  0.003234   \n",
       "3           6          1             219           225   GTCATTG -0.009522   \n",
       "4           7          1             226           233  AATTCTAC -0.004358   \n",
       "...       ...        ...             ...           ...       ...       ...   \n",
       "94483  121154         20             410           415    TAACTT  0.017504   \n",
       "94484  121155         20             416           420     TTTTA  0.001789   \n",
       "94485  121156         20             421           427   TCTTTAT  0.003276   \n",
       "94486  121157         20             428           432     CCCAC  0.005503   \n",
       "94487  121159         20             437           443   TAAATTT  0.009119   \n",
       "\n",
       "      feature_type            id                               name  gene  \\\n",
       "0              CDS  CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "1              CDS  CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "2              CDS  CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "3              CDS  CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "4              CDS  CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "...            ...           ...                                ...   ...   \n",
       "94483          CDS  KKIOAG_13685     YopX domain-containing protein  yopx   \n",
       "94484          CDS  KKIOAG_13685     YopX domain-containing protein  yopx   \n",
       "94485          CDS  KKIOAG_13685     YopX domain-containing protein  yopx   \n",
       "94486          CDS  KKIOAG_13685     YopX domain-containing protein  yopx   \n",
       "94487          CDS  KKIOAG_13685     YopX domain-containing protein  yopx   \n",
       "\n",
       "                                               atributes tube_code  person_id  \\\n",
       "0      ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065    2221447   \n",
       "1      ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065    2221447   \n",
       "2      ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065    2221447   \n",
       "3      ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065    2221447   \n",
       "4      ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065    2221447   \n",
       "...                                                  ...       ...        ...   \n",
       "94483  ID=KKIOAG_13685;Name=YopX domain-containing pr...  AH21K020    2035444   \n",
       "94484  ID=KKIOAG_13685;Name=YopX domain-containing pr...  AH21K020    2035444   \n",
       "94485  ID=KKIOAG_13685;Name=YopX domain-containing pr...  AH21K020    2035444   \n",
       "94486  ID=KKIOAG_13685;Name=YopX domain-containing pr...  AH21K020    2035444   \n",
       "94487  ID=KKIOAG_13685;Name=YopX domain-containing pr...  AH21K020    2035444   \n",
       "\n",
       "       JOURNEY_ID  \n",
       "0          681015  \n",
       "1          681015  \n",
       "2          681015  \n",
       "3          681015  \n",
       "4          681015  \n",
       "...           ...  \n",
       "94483    15586881  \n",
       "94484    15586881  \n",
       "94485    15586881  \n",
       "94486    15586881  \n",
       "94487    15586881  \n",
       "\n",
       "[73198 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "overlappingFilesDir = Path(os.environ['GENOMICS_DATA_BASE'], 'genome_nlp_tokens', 'overlapping_with_annotations')\n",
    "\n",
    "overlappingDfList = []\n",
    "for overlappingFile in os.listdir(overlappingFilesDir):\n",
    "    df = pd.read_csv(Path(overlappingFilesDir, overlappingFile), sep='\\t', names=['contig_id', 'start_position', 'end_position', 'tokens', 'score', 'feature_type', 'id', 'name', 'gene', 'atributes'])\n",
    "    df['tube_code'] = [overlappingFile.split('_')[0]]*df.shape[0]\n",
    "    overlappingDfList.append(df)\n",
    "overlappingDf = pd.concat(overlappingDfList, ignore_index=True)\n",
    "overlappingDf = overlappingDf.merge(\n",
    "    mappedJourneyTubeIdsDf,\n",
    "    how='inner',\n",
    "    on=['tube_code']\n",
    ")\n",
    "overlappingDf = overlappingDf.drop_duplicates()\n",
    "overlappingDf = overlappingDf[overlappingDf.tokens.apply(lambda x: (len(x) >= 5))].reset_index()\n",
    "overlappingDf['gene'] = overlappingDf.gene.str.lower()\n",
    "overlappingDf = overlappingDf[overlappingDf.feature_type.isin(['CDS', 'ncRNA', 'oriC', 'regulatory_region', 'oriT'])]\n",
    "overlappingDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.034861431584150615)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff = overlappingDf.score.mean() + 2 * overlappingDf.score.std()\n",
    "cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52.50000000000001 0.2061361964240253 False\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`The test indicates that the tokens associated with the high score group are significantly different than the rest`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1639017/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1639017/2494458605.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CACCGT</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>0.004666</td>\n",
       "      <td>1.348462</td>\n",
       "      <td>4.449512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ATGAAA</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>0.004666</td>\n",
       "      <td>1.348462</td>\n",
       "      <td>5.191098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GAGCA</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>0.003181</td>\n",
       "      <td>0.919406</td>\n",
       "      <td>6.525952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTTTG</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>0.004454</td>\n",
       "      <td>1.287169</td>\n",
       "      <td>7.768990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CACTAT</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>0.002333</td>\n",
       "      <td>0.674231</td>\n",
       "      <td>11.865366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAAGG</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001273</td>\n",
       "      <td>0.367762</td>\n",
       "      <td>24.472318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTCTTTGT</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>0.306469</td>\n",
       "      <td>29.366782</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "7    CACCGT         6        22       0.004666        1.348462   \n",
       "5    ATGAAA         7        22       0.004666        1.348462   \n",
       "8     GAGCA         6        15       0.003181        0.919406   \n",
       "0     CTTTG        10        21       0.004454        1.287169   \n",
       "4    CACTAT         8        11       0.002333        0.674231   \n",
       "2    ACAAGG         9         6       0.001273        0.367762   \n",
       "3  TTCTTTGT         9         5       0.001060        0.306469   \n",
       "\n",
       "   ratio_difference  \n",
       "7          4.449512  \n",
       "5          5.191098  \n",
       "8          6.525952  \n",
       "0          7.768990  \n",
       "4         11.865366  \n",
       "2         24.472318  \n",
       "3         29.366782  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_type</th>\n",
       "      <th>gene</th>\n",
       "      <th>name</th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>esad</td>\n",
       "      <td>type VII secretion system secreted protein%2C ...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CDS</td>\n",
       "      <td>rplq</td>\n",
       "      <td>50S ribosomal protein L17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CDS</td>\n",
       "      <td>tnp</td>\n",
       "      <td>IS3 family transposase ORF B</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDS</td>\n",
       "      <td>fni</td>\n",
       "      <td>type 2 isopentenyl-diphosphate Delta-isomerase</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CDS</td>\n",
       "      <td>noc</td>\n",
       "      <td>Chromosome segregation protein Spo0J%2C contai...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CDS</td>\n",
       "      <td>tagb</td>\n",
       "      <td>CDP-glycerol glycerophosphotransferase%2C TagB...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CDS</td>\n",
       "      <td>tnp</td>\n",
       "      <td>IS3 family ISSau2 transposase ORF A</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>rnaiii</td>\n",
       "      <td>(3' truncated) RNAIII</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CDS</td>\n",
       "      <td>trka</td>\n",
       "      <td>Trk/Ktr K+ transport system regulatory compone...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>muts</td>\n",
       "      <td>DNA mismatch repair ATPase MutS</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CDS</td>\n",
       "      <td>pknb</td>\n",
       "      <td>serine/threonine protein kinase Stk1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>crr</td>\n",
       "      <td>PTS system glucose-specific EIIA component</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>deob</td>\n",
       "      <td>phosphopentomutase</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CDS</td>\n",
       "      <td>yopx</td>\n",
       "      <td>YopX domain-containing protein</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CDS</td>\n",
       "      <td>sea</td>\n",
       "      <td>staphylococcal enterotoxin type A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   feature_type    gene                                               name  \\\n",
       "3           CDS    esad  type VII secretion system secreted protein%2C ...   \n",
       "18          CDS    rplq                          50S ribosomal protein L17   \n",
       "22          CDS     tnp                       IS3 family transposase ORF B   \n",
       "4           CDS     fni     type 2 isopentenyl-diphosphate Delta-isomerase   \n",
       "12          CDS     noc  Chromosome segregation protein Spo0J%2C contai...   \n",
       "20          CDS    tagb  CDP-glycerol glycerophosphotransferase%2C TagB...   \n",
       "21          CDS     tnp                IS3 family ISSau2 transposase ORF A   \n",
       "26        ncRNA  rnaiii                              (3' truncated) RNAIII   \n",
       "23          CDS    trka  Trk/Ktr K+ transport system regulatory compone...   \n",
       "11          CDS    muts                    DNA mismatch repair ATPase MutS   \n",
       "14          CDS    pknb               serine/threonine protein kinase Stk1   \n",
       "1           CDS     crr         PTS system glucose-specific EIIA component   \n",
       "2           CDS    deob                                 phosphopentomutase   \n",
       "25          CDS    yopx                     YopX domain-containing protein   \n",
       "19          CDS     sea                  staphylococcal enterotoxin type A   \n",
       "\n",
       "    tokens  \n",
       "3       27  \n",
       "18       9  \n",
       "22       8  \n",
       "4        4  \n",
       "12       3  \n",
       "20       3  \n",
       "21       3  \n",
       "26       3  \n",
       "23       2  \n",
       "11       2  \n",
       "14       2  \n",
       "1        2  \n",
       "2        2  \n",
       "25       2  \n",
       "19       2  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tempDf = overrepresentedTokensDf.merge(overlappingDf[['tokens', 'feature_type', 'id', 'name', 'gene']], how='inner', on=['tokens'])\n",
    "\n",
    "tempDf[['tokens', 'feature_type', 'gene', 'name']].groupby(by=['feature_type', 'gene', 'name']).agg('count').reset_index().sort_values(by=['tokens'], ascending=False)[:15]\n",
    "# gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>type VII secretion system secreted protein%2C ...</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DUF600 domain-containing protein</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50S ribosomal protein L17</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IS3 family transposase ORF B</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DUF5079 domain-containing protein</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0  type VII secretion system secreted protein%2C ...     27\n",
       "1                   DUF600 domain-containing protein     12\n",
       "2                          50S ribosomal protein L17      9\n",
       "3                       IS3 family transposase ORF B      8\n",
       "4                  DUF5079 domain-containing protein      6"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1638799/3615638238.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AAATAT</td>\n",
       "      <td>5</td>\n",
       "      <td>100</td>\n",
       "      <td>0.021209</td>\n",
       "      <td>6.129374</td>\n",
       "      <td>0.815744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ATGGT</td>\n",
       "      <td>5</td>\n",
       "      <td>89</td>\n",
       "      <td>0.018876</td>\n",
       "      <td>5.455143</td>\n",
       "      <td>0.916566</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "10  AAATAT         5       100       0.021209        6.129374   \n",
       "11   ATGGT         5        89       0.018876        5.455143   \n",
       "\n",
       "    ratio_difference  \n",
       "10          0.815744  \n",
       "11          0.916566  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 1]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ilva</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hly</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>esag</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ara1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tnp</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0  ilva     21\n",
       "1   hly     10\n",
       "2  esag      8\n",
       "3  ara1      7\n",
       "4   tnp      6"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>threonine ammonia-lyase IlvA</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alpha-Hemolysin precursor</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TIGR01741 family protein</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Uncharacterized membrane-anchored protein</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aldo/keto reductase%2C related to diketogulona...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0                       threonine ammonia-lyase IlvA     21\n",
       "1                          Alpha-Hemolysin precursor     10\n",
       "2                           TIGR01741 family protein      9\n",
       "3          Uncharacterized membrane-anchored protein      8\n",
       "4  Aldo/keto reductase%2C related to diketogulona...      7"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(-0.04383813452765485)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff = overlappingDf.score.mean() - 2 * overlappingDf.score.std()\n",
    "cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1282.0340721733464 5.533405826241168e-05 True\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1638799/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1638799/1066762906.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 2]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ATCGC</td>\n",
       "      <td>151</td>\n",
       "      <td>5</td>\n",
       "      <td>0.001870</td>\n",
       "      <td>58.679880</td>\n",
       "      <td>2.573284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATGAT</td>\n",
       "      <td>297</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>105.623785</td>\n",
       "      <td>2.811867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACGAT</td>\n",
       "      <td>254</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002244</td>\n",
       "      <td>70.415856</td>\n",
       "      <td>3.607142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tokens  count_hs  count_ls  proportion_ls  count_expected  ratio_difference\n",
       "11  ATCGC       151         5       0.001870       58.679880          2.573284\n",
       "0   ATGAT       297         9       0.003366      105.623785          2.811867\n",
       "1   ACGAT       254         6       0.002244       70.415856          3.607142"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 2]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tnp</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mrp</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aro8</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>eap</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>esag</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0   tnp     40\n",
       "1   mrp     36\n",
       "2  aro8     21\n",
       "3   eap     12\n",
       "4  esag     10"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Fe-S cluster carrier ATPase%2C Mrp/ApbC/NBP35 ...</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Phage protein</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA-binding transcriptional regulator%2C MocR ...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UPF0344 protein SAR0931</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TIGR01741 family protein</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0  Fe-S cluster carrier ATPase%2C Mrp/ApbC/NBP35 ...     36\n",
       "1                                      Phage protein     29\n",
       "2  DNA-binding transcriptional regulator%2C MocR ...     21\n",
       "3                            UPF0344 protein SAR0931     20\n",
       "4                           TIGR01741 family protein     17"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1638799/3318705502.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td>TAGTTTTT</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>105.623785</td>\n",
       "      <td>0.056805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>CTTGAT</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>0.005610</td>\n",
       "      <td>176.039641</td>\n",
       "      <td>0.068166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764</th>\n",
       "      <td>ATATATTT</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0.004114</td>\n",
       "      <td>129.095737</td>\n",
       "      <td>0.085208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>GACCTC</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>105.623785</td>\n",
       "      <td>0.085208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891</th>\n",
       "      <td>CACTGC</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002244</td>\n",
       "      <td>70.415856</td>\n",
       "      <td>0.085208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>TGCCTT</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>105.623785</td>\n",
       "      <td>0.085208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>GGTTG</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>105.623785</td>\n",
       "      <td>0.094676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>CTAAGT</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002244</td>\n",
       "      <td>70.415856</td>\n",
       "      <td>0.099409</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "892  TAGTTTTT         6         9       0.003366      105.623785   \n",
       "741    CTTGAT        12        15       0.005610      176.039641   \n",
       "764  ATATATTT        11        11       0.004114      129.095737   \n",
       "817    GACCTC         9         9       0.003366      105.623785   \n",
       "891    CACTGC         6         6       0.002244       70.415856   \n",
       "833    TGCCTT         9         9       0.003366      105.623785   \n",
       "812     GGTTG        10         9       0.003366      105.623785   \n",
       "869    CTAAGT         7         6       0.002244       70.415856   \n",
       "\n",
       "     ratio_difference  \n",
       "892          0.056805  \n",
       "741          0.068166  \n",
       "764          0.085208  \n",
       "817          0.085208  \n",
       "891          0.085208  \n",
       "833          0.085208  \n",
       "812          0.094676  \n",
       "869          0.099409  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.1]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>esad</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rplq</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>splc</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aro8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>meca</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0  esad     54\n",
       "1  rplq     14\n",
       "2  splc      5\n",
       "3  aro8      4\n",
       "4  meca      4"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>type VII secretion system secreted protein%2C ...</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>50S ribosomal protein L17</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>serine protease SplC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phage protein</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNA-binding transcriptional regulator%2C MocR ...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0  type VII secretion system secreted protein%2C ...     54\n",
       "1                          50S ribosomal protein L17     14\n",
       "2                               serine protease SplC      5\n",
       "3                                      Phage protein      4\n",
       "4  DNA-binding transcriptional regulator%2C MocR ...      4"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
