{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistical analysis to get the significantly associated tokens for attribution score groupings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read mapping information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>visit_occurrence_id</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>SEX</th>\n",
       "      <th>Age_Bin</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>678487</td>\n",
       "      <td>3312694</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>3312694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2094045</td>\n",
       "      <td>9199249</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(20, 30]</td>\n",
       "      <td>9199249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2107813</td>\n",
       "      <td>9410987</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(50, 60]</td>\n",
       "      <td>9410987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2083212</td>\n",
       "      <td>602890</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2083212</td>\n",
       "      <td>9039437</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7194</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16588204</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16588204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7195</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16709341</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16709341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7196</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16651755</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16651755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7197</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16872774</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16872774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16869916</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16869916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7199 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      person_id  visit_occurrence_id       TYPE COUNTRY_OF_BIRTH     SEX  \\\n",
       "0        678487              3312694  Emergency        Australia    Male   \n",
       "1       2094045              9199249  Emergency        Australia    Male   \n",
       "2       2107813              9410987  Emergency        Australia    Male   \n",
       "3       2083212               602890  Inpatient        Australia  Female   \n",
       "4       2083212              9039437  Emergency        Australia  Female   \n",
       "...         ...                  ...        ...              ...     ...   \n",
       "7194    2603535             16588204  Inpatient        Australia    Male   \n",
       "7195    2603535             16709341  Inpatient        Australia    Male   \n",
       "7196    2603535             16651755  Inpatient        Australia    Male   \n",
       "7197    2098616             16872774  Emergency        Australia  Female   \n",
       "7198    2098616             16869916  Emergency        Australia  Female   \n",
       "\n",
       "       Age_Bin  JOURNEY_ID  \n",
       "0     (60, 70]     3312694  \n",
       "1     (20, 30]     9199249  \n",
       "2     (50, 60]     9410987  \n",
       "3     (40, 50]     9039437  \n",
       "4     (40, 50]     9039437  \n",
       "...        ...         ...  \n",
       "7194  (60, 70]    16588204  \n",
       "7195  (60, 70]    16709341  \n",
       "7196  (60, 70]    16651755  \n",
       "7197  (80, 90]    16872774  \n",
       "7198  (80, 90]    16869916  \n",
       "\n",
       "[7199 rows x 7 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "demoTrainDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_train.csv', sep='\\t')\n",
    "demoTestDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_test.csv', sep='\\t')\n",
    "demoValidateDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_validate.csv', sep='\\t')\n",
    "demoDf = pd.concat([demoTrainDf, demoTestDf, demoValidateDf], ignore_index=True)\n",
    "demoDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tube_code</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>db_ID</th>\n",
       "      <th>pt_age</th>\n",
       "      <th>pt_gender</th>\n",
       "      <th>EPISODE_ID</th>\n",
       "      <th>hospital_admission</th>\n",
       "      <th>hospital_discharge</th>\n",
       "      <th>hospital_in_last_year</th>\n",
       "      <th>genome_species</th>\n",
       "      <th>species_reported</th>\n",
       "      <th>contig_number</th>\n",
       "      <th>length</th>\n",
       "      <th>tube_code_duplicate_or_old</th>\n",
       "      <th>date_of_collection</th>\n",
       "      <th>ID_number</th>\n",
       "      <th>location_additional_02</th>\n",
       "      <th>collected_from_original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH19J072</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>92</td>\n",
       "      <td>5131021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-0997</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH19J074</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus sp. D19</td>\n",
       "      <td>Streptococcus mitis group</td>\n",
       "      <td>84</td>\n",
       "      <td>2022991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-1174</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH19B003</td>\n",
       "      <td>526238.0</td>\n",
       "      <td>567AE</td>\n",
       "      <td>85</td>\n",
       "      <td>Male</td>\n",
       "      <td>640261.0</td>\n",
       "      <td>2019-02-04</td>\n",
       "      <td>2019-02-07</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>93</td>\n",
       "      <td>4942915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-02-03</td>\n",
       "      <td>19-034-1596</td>\n",
       "      <td>A-4WB;GMC</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH20A024</td>\n",
       "      <td>788941.0</td>\n",
       "      <td>3YBDS</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>13194824.0</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>110</td>\n",
       "      <td>5149157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>20-016-2112</td>\n",
       "      <td>S-ED;Emergency</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH20L041</td>\n",
       "      <td>1023735.0</td>\n",
       "      <td>2VJ78</td>\n",
       "      <td>42</td>\n",
       "      <td>Male</td>\n",
       "      <td>14210147.0</td>\n",
       "      <td>2020-12-18</td>\n",
       "      <td>2021-01-05</td>\n",
       "      <td>yes</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>46</td>\n",
       "      <td>2735216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-16</td>\n",
       "      <td>20-351-2856</td>\n",
       "      <td>Emergency Dept;nan</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3112</th>\n",
       "      <td>ALF22L076</td>\n",
       "      <td>639020.0</td>\n",
       "      <td>3S6HL</td>\n",
       "      <td>84</td>\n",
       "      <td>Male</td>\n",
       "      <td>17219899.0</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-20</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus vestibularis</td>\n",
       "      <td>Streptococcus salivarius group</td>\n",
       "      <td>107</td>\n",
       "      <td>1938627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>22-349-2150</td>\n",
       "      <td>Ward 4WB</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>ALF22L089</td>\n",
       "      <td>655404.0</td>\n",
       "      <td>6EWBM</td>\n",
       "      <td>67</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus columbae</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>152</td>\n",
       "      <td>3972219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-17</td>\n",
       "      <td>22-351-1465</td>\n",
       "      <td>Comm/Amb Clinic</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3114</th>\n",
       "      <td>ALF22L138</td>\n",
       "      <td>2572048.0</td>\n",
       "      <td>YM5PJ</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter cloacae complex</td>\n",
       "      <td>66</td>\n",
       "      <td>4876179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-27</td>\n",
       "      <td>22-361-0202</td>\n",
       "      <td>Ward 7 East</td>\n",
       "      <td>Hick White Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3115</th>\n",
       "      <td>ALF23A102</td>\n",
       "      <td>2054964.0</td>\n",
       "      <td>5DYAH</td>\n",
       "      <td>35</td>\n",
       "      <td>Female</td>\n",
       "      <td>17294287.0</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus constellatus</td>\n",
       "      <td>Streptococcus anginosus group</td>\n",
       "      <td>54</td>\n",
       "      <td>1869828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-11</td>\n",
       "      <td>23-011-3305</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>ALF23B102</td>\n",
       "      <td>2015952.0</td>\n",
       "      <td>57E2W</td>\n",
       "      <td>82</td>\n",
       "      <td>Female</td>\n",
       "      <td>17281239.0</td>\n",
       "      <td>2023-01-08</td>\n",
       "      <td>2023-02-16</td>\n",
       "      <td>no</td>\n",
       "      <td>Acinetobacter colistiniresistens</td>\n",
       "      <td>Acinetobacter species</td>\n",
       "      <td>233</td>\n",
       "      <td>3952992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-15</td>\n",
       "      <td>23-046-3310</td>\n",
       "      <td>ICU</td>\n",
       "      <td>PICC Line White</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3117 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      tube_code  PATIENT_ID  db_ID  pt_age pt_gender  EPISODE_ID  \\\n",
       "0      AH19J072    213972.0  3JNQS      64      Male  12931474.0   \n",
       "1      AH19J074    213972.0  3JNQS      64      Male  12931474.0   \n",
       "2      AH19B003    526238.0  567AE      85      Male    640261.0   \n",
       "3      AH20A024    788941.0  3YBDS      58    Female  13194824.0   \n",
       "4      AH20L041   1023735.0  2VJ78      42      Male  14210147.0   \n",
       "...         ...         ...    ...     ...       ...         ...   \n",
       "3112  ALF22L076    639020.0  3S6HL      84      Male  17219899.0   \n",
       "3113  ALF22L089    655404.0  6EWBM      67    Female         NaN   \n",
       "3114  ALF22L138   2572048.0  YM5PJ      61      Male         NaN   \n",
       "3115  ALF23A102   2054964.0  5DYAH      35    Female  17294287.0   \n",
       "3116  ALF23B102   2015952.0  57E2W      82    Female  17281239.0   \n",
       "\n",
       "     hospital_admission hospital_discharge hospital_in_last_year  \\\n",
       "0            2019-10-17         2019-11-29                   yes   \n",
       "1            2019-10-17         2019-11-29                   yes   \n",
       "2            2019-02-04         2019-02-07                    no   \n",
       "3            2020-01-16         2020-01-19                    no   \n",
       "4            2020-12-18         2021-01-05                   yes   \n",
       "...                 ...                ...                   ...   \n",
       "3112         2022-12-15         2022-12-20                   yes   \n",
       "3113                NaN                NaN                   NaN   \n",
       "3114                NaN                NaN                   NaN   \n",
       "3115         2023-01-12         2023-02-01                   yes   \n",
       "3116         2023-01-08         2023-02-16                    no   \n",
       "\n",
       "                        genome_species                species_reported  \\\n",
       "0                     Escherichia coli                Escherichia coli   \n",
       "1                Streptococcus sp. D19       Streptococcus mitis group   \n",
       "2                     Escherichia coli                Escherichia coli   \n",
       "3                     Escherichia coli                Escherichia coli   \n",
       "4                Staphylococcus aureus           Staphylococcus aureus   \n",
       "...                                ...                             ...   \n",
       "3112        Streptococcus vestibularis  Streptococcus salivarius group   \n",
       "3113                  Proteus columbae                Proteus vulgaris   \n",
       "3114         Enterobacter roggenkampii    Enterobacter cloacae complex   \n",
       "3115        Streptococcus constellatus   Streptococcus anginosus group   \n",
       "3116  Acinetobacter colistiniresistens           Acinetobacter species   \n",
       "\n",
       "      contig_number   length tube_code_duplicate_or_old date_of_collection  \\\n",
       "0                92  5131021                        NaN         2019-10-30   \n",
       "1                84  2022991                        NaN         2019-10-30   \n",
       "2                93  4942915                        NaN         2019-02-03   \n",
       "3               110  5149157                        NaN         2020-01-16   \n",
       "4                46  2735216                        NaN         2020-12-16   \n",
       "...             ...      ...                        ...                ...   \n",
       "3112            107  1938627                        NaN         2022-12-15   \n",
       "3113            152  3972219                        NaN         2022-12-17   \n",
       "3114             66  4876179                        NaN         2022-12-27   \n",
       "3115             54  1869828                        NaN         2023-01-11   \n",
       "3116            233  3952992                        NaN         2023-02-15   \n",
       "\n",
       "        ID_number location_additional_02 collected_from_original  \n",
       "0     19-303-0997             A-7EA;HAEM                   Blood  \n",
       "1     19-303-1174             A-7EA;HAEM        Blood Peripheral  \n",
       "2     19-034-1596              A-4WB;GMC                   Blood  \n",
       "3     20-016-2112         S-ED;Emergency            Blood Venous  \n",
       "4     20-351-2856     Emergency Dept;nan                   Blood  \n",
       "...           ...                    ...                     ...  \n",
       "3112  22-349-2150               Ward 4WB        Blood Peripheral  \n",
       "3113  22-351-1465        Comm/Amb Clinic                   Blood  \n",
       "3114  22-361-0202            Ward 7 East          Hick White Lum  \n",
       "3115  23-011-3305                    ICU            Blood Venous  \n",
       "3116  23-046-3310                    ICU         PICC Line White  \n",
       "\n",
       "[3117 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/patient_tube_id_mapping_full.tsv', sep='\\t')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "      <th>tube_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50056</td>\n",
       "      <td>13224842</td>\n",
       "      <td>AH20B011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62567</td>\n",
       "      <td>12645629</td>\n",
       "      <td>AH19G065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138108</td>\n",
       "      <td>14737781</td>\n",
       "      <td>AH21E052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>AH20J009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>KPN2214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>2084373</td>\n",
       "      <td>16158768</td>\n",
       "      <td>ALF22C004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>1058066</td>\n",
       "      <td>16759482</td>\n",
       "      <td>ALF22H089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>2552781</td>\n",
       "      <td>16966609</td>\n",
       "      <td>ALF22J054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2638484</td>\n",
       "      <td>16969999</td>\n",
       "      <td>ALF22K143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>2015952</td>\n",
       "      <td>17281239</td>\n",
       "      <td>ALF23B102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>517 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     person_id  JOURNEY_ID  tube_code\n",
       "0        50056    13224842   AH20B011\n",
       "1        62567    12645629   AH19G065\n",
       "2       138108    14737781   AH21E052\n",
       "3       211314    13955099   AH20J009\n",
       "4       211314    13955099    KPN2214\n",
       "..         ...         ...        ...\n",
       "512    2084373    16158768  ALF22C004\n",
       "513    1058066    16759482  ALF22H089\n",
       "514    2552781    16966609  ALF22J054\n",
       "515    2638484    16969999  ALF22K143\n",
       "516    2015952    17281239  ALF23B102\n",
       "\n",
       "[517 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mappedJourneyTubeIdsDf = mappingDf[['tube_code', 'PATIENT_ID', 'EPISODE_ID']].drop_duplicates().merge(\n",
    "    demoDf[['person_id', 'visit_occurrence_id', 'JOURNEY_ID']],\n",
    "    how='inner',\n",
    "    left_on=['PATIENT_ID', 'EPISODE_ID'],\n",
    "    right_on=['person_id', 'visit_occurrence_id']\n",
    ").drop(\n",
    "    columns=['PATIENT_ID', 'EPISODE_ID', 'visit_occurrence_id']\n",
    ")[['person_id', 'JOURNEY_ID', 'tube_code']].drop_duplicates()\n",
    "mappedJourneyTubeIdsDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read bed files containing overlapping tokens and attribution scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>tokens</th>\n",
       "      <th>score</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>atributes</th>\n",
       "      <th>tube_code</th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>214</td>\n",
       "      <td>219</td>\n",
       "      <td>TGAGGG</td>\n",
       "      <td>-0.000951</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CENNFK_00005</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...</td>\n",
       "      <td>AH21B039</td>\n",
       "      <td>908851</td>\n",
       "      <td>14408941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>220</td>\n",
       "      <td>224</td>\n",
       "      <td>GTTAT</td>\n",
       "      <td>-0.000473</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CENNFK_00005</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...</td>\n",
       "      <td>AH21B039</td>\n",
       "      <td>908851</td>\n",
       "      <td>14408941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>228</td>\n",
       "      <td>233</td>\n",
       "      <td>GAAAGA</td>\n",
       "      <td>-0.000867</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CENNFK_00005</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...</td>\n",
       "      <td>AH21B039</td>\n",
       "      <td>908851</td>\n",
       "      <td>14408941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>234</td>\n",
       "      <td>239</td>\n",
       "      <td>TATTGT</td>\n",
       "      <td>-0.000915</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CENNFK_00005</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...</td>\n",
       "      <td>AH21B039</td>\n",
       "      <td>908851</td>\n",
       "      <td>14408941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>240</td>\n",
       "      <td>245</td>\n",
       "      <td>AGCTGG</td>\n",
       "      <td>-0.001374</td>\n",
       "      <td>CDS</td>\n",
       "      <td>CENNFK_00005</td>\n",
       "      <td>Transposase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...</td>\n",
       "      <td>AH21B039</td>\n",
       "      <td>908851</td>\n",
       "      <td>14408941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878529</th>\n",
       "      <td>18368428</td>\n",
       "      <td>5</td>\n",
       "      <td>205763</td>\n",
       "      <td>205769</td>\n",
       "      <td>ATGATAT</td>\n",
       "      <td>0.002546</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_10525</td>\n",
       "      <td>Prepilin signal peptidase PulO (type II secret...</td>\n",
       "      <td>pulo</td>\n",
       "      <td>ID=KKIOAG_10525;Name=Prepilin signal peptidase...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878530</th>\n",
       "      <td>18368429</td>\n",
       "      <td>5</td>\n",
       "      <td>205770</td>\n",
       "      <td>205775</td>\n",
       "      <td>CCATCA</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_10525</td>\n",
       "      <td>Prepilin signal peptidase PulO (type II secret...</td>\n",
       "      <td>pulo</td>\n",
       "      <td>ID=KKIOAG_10525;Name=Prepilin signal peptidase...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878531</th>\n",
       "      <td>18368430</td>\n",
       "      <td>5</td>\n",
       "      <td>205776</td>\n",
       "      <td>205781</td>\n",
       "      <td>ATTTTT</td>\n",
       "      <td>-0.002056</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_10525</td>\n",
       "      <td>Prepilin signal peptidase PulO (type II secret...</td>\n",
       "      <td>pulo</td>\n",
       "      <td>ID=KKIOAG_10525;Name=Prepilin signal peptidase...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878532</th>\n",
       "      <td>18368431</td>\n",
       "      <td>5</td>\n",
       "      <td>205782</td>\n",
       "      <td>205787</td>\n",
       "      <td>AGGAGG</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_10525</td>\n",
       "      <td>Prepilin signal peptidase PulO (type II secret...</td>\n",
       "      <td>pulo</td>\n",
       "      <td>ID=KKIOAG_10525;Name=Prepilin signal peptidase...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14878533</th>\n",
       "      <td>18368433</td>\n",
       "      <td>5</td>\n",
       "      <td>205791</td>\n",
       "      <td>205798</td>\n",
       "      <td>ATATTTTT</td>\n",
       "      <td>-0.000752</td>\n",
       "      <td>CDS</td>\n",
       "      <td>KKIOAG_10525</td>\n",
       "      <td>Prepilin signal peptidase PulO (type II secret...</td>\n",
       "      <td>pulo</td>\n",
       "      <td>ID=KKIOAG_10525;Name=Prepilin signal peptidase...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14870257 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             index  contig_id  start_position  end_position    tokens  \\\n",
       "0                0          1             214           219    TGAGGG   \n",
       "1                1          1             220           224     GTTAT   \n",
       "2                3          1             228           233    GAAAGA   \n",
       "3                4          1             234           239    TATTGT   \n",
       "4                5          1             240           245    AGCTGG   \n",
       "...            ...        ...             ...           ...       ...   \n",
       "14878529  18368428          5          205763        205769   ATGATAT   \n",
       "14878530  18368429          5          205770        205775    CCATCA   \n",
       "14878531  18368430          5          205776        205781    ATTTTT   \n",
       "14878532  18368431          5          205782        205787    AGGAGG   \n",
       "14878533  18368433          5          205791        205798  ATATTTTT   \n",
       "\n",
       "             score feature_type            id  \\\n",
       "0        -0.000951          CDS  CENNFK_00005   \n",
       "1        -0.000473          CDS  CENNFK_00005   \n",
       "2        -0.000867          CDS  CENNFK_00005   \n",
       "3        -0.000915          CDS  CENNFK_00005   \n",
       "4        -0.001374          CDS  CENNFK_00005   \n",
       "...            ...          ...           ...   \n",
       "14878529  0.002546          CDS  KKIOAG_10525   \n",
       "14878530  0.000022          CDS  KKIOAG_10525   \n",
       "14878531 -0.002056          CDS  KKIOAG_10525   \n",
       "14878532  0.000008          CDS  KKIOAG_10525   \n",
       "14878533 -0.000752          CDS  KKIOAG_10525   \n",
       "\n",
       "                                                       name  gene  \\\n",
       "0                                               Transposase   NaN   \n",
       "1                                               Transposase   NaN   \n",
       "2                                               Transposase   NaN   \n",
       "3                                               Transposase   NaN   \n",
       "4                                               Transposase   NaN   \n",
       "...                                                     ...   ...   \n",
       "14878529  Prepilin signal peptidase PulO (type II secret...  pulo   \n",
       "14878530  Prepilin signal peptidase PulO (type II secret...  pulo   \n",
       "14878531  Prepilin signal peptidase PulO (type II secret...  pulo   \n",
       "14878532  Prepilin signal peptidase PulO (type II secret...  pulo   \n",
       "14878533  Prepilin signal peptidase PulO (type II secret...  pulo   \n",
       "\n",
       "                                                  atributes tube_code  \\\n",
       "0         ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...  AH21B039   \n",
       "1         ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...  AH21B039   \n",
       "2         ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...  AH21B039   \n",
       "3         ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...  AH21B039   \n",
       "4         ID=CENNFK_00005;Name=Transposase;locus_tag=CEN...  AH21B039   \n",
       "...                                                     ...       ...   \n",
       "14878529  ID=KKIOAG_10525;Name=Prepilin signal peptidase...  AH21K020   \n",
       "14878530  ID=KKIOAG_10525;Name=Prepilin signal peptidase...  AH21K020   \n",
       "14878531  ID=KKIOAG_10525;Name=Prepilin signal peptidase...  AH21K020   \n",
       "14878532  ID=KKIOAG_10525;Name=Prepilin signal peptidase...  AH21K020   \n",
       "14878533  ID=KKIOAG_10525;Name=Prepilin signal peptidase...  AH21K020   \n",
       "\n",
       "          person_id  JOURNEY_ID  \n",
       "0            908851    14408941  \n",
       "1            908851    14408941  \n",
       "2            908851    14408941  \n",
       "3            908851    14408941  \n",
       "4            908851    14408941  \n",
       "...             ...         ...  \n",
       "14878529    2035444    15586881  \n",
       "14878530    2035444    15586881  \n",
       "14878531    2035444    15586881  \n",
       "14878532    2035444    15586881  \n",
       "14878533    2035444    15586881  \n",
       "\n",
       "[14870257 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "overlappingFilesDir = Path(os.environ['GENOMICS_DATA_BASE'], 'genome_nlp_tokens', 'overlapping_with_annotations')\n",
    "\n",
    "overlappingDfList = []\n",
    "for overlappingFile in os.listdir(overlappingFilesDir):\n",
    "    df = pd.read_csv(Path(overlappingFilesDir, overlappingFile), sep='\\t', names=['contig_id', 'start_position', 'end_position', 'tokens', 'score', 'feature_type', 'id', 'name', 'gene', 'atributes'])\n",
    "    df['tube_code'] = [overlappingFile.split('_')[0]]*df.shape[0]\n",
    "    overlappingDfList.append(df)\n",
    "overlappingDf = pd.concat(overlappingDfList, ignore_index=True)\n",
    "overlappingDf = overlappingDf.merge(\n",
    "    mappedJourneyTubeIdsDf,\n",
    "    how='inner',\n",
    "    on=['tube_code']\n",
    ")\n",
    "overlappingDf = overlappingDf.drop_duplicates()\n",
    "overlappingDf = overlappingDf[overlappingDf.tokens.apply(lambda x: (len(x) >= 5))].reset_index()\n",
    "overlappingDf['gene'] = overlappingDf.gene.str.lower()\n",
    "overlappingDf = overlappingDf[overlappingDf.feature_type.isin(['CDS', 'ncRNA', 'oriC', 'regulatory_region', 'oriT'])]\n",
    "overlappingDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.003378765748551873)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff = overlappingDf.score.mean() + 1 * overlappingDf.score.std()\n",
    "cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "727805.0274963141 3.55127562151092e-20 True\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`The test indicates that the tokens associated with the high score group are significantly different than the rest`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/1219838856.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 10]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AAAATGT</td>\n",
       "      <td>1249</td>\n",
       "      <td>7343</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>123.669344</td>\n",
       "      <td>10.099512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>CTGTGG</td>\n",
       "      <td>89</td>\n",
       "      <td>512</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>8.623002</td>\n",
       "      <td>10.321231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>TCAGTAT</td>\n",
       "      <td>410</td>\n",
       "      <td>2348</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>39.544548</td>\n",
       "      <td>10.368054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>AGATGG</td>\n",
       "      <td>389</td>\n",
       "      <td>2227</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>37.506691</td>\n",
       "      <td>10.371483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTGCAA</td>\n",
       "      <td>1203</td>\n",
       "      <td>6755</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>113.766365</td>\n",
       "      <td>10.574303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAATTT</td>\n",
       "      <td>1955</td>\n",
       "      <td>10968</td>\n",
       "      <td>0.000847</td>\n",
       "      <td>184.720872</td>\n",
       "      <td>10.583536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>TTATATA</td>\n",
       "      <td>354</td>\n",
       "      <td>1983</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>33.397291</td>\n",
       "      <td>10.599662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>AAGAAAAAAA</td>\n",
       "      <td>55</td>\n",
       "      <td>307</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>5.170433</td>\n",
       "      <td>10.637407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>TCCTTTCC</td>\n",
       "      <td>30</td>\n",
       "      <td>167</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>2.812581</td>\n",
       "      <td>10.666360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CTGACT</td>\n",
       "      <td>393</td>\n",
       "      <td>2104</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>35.435149</td>\n",
       "      <td>11.090683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>GCTCAG</td>\n",
       "      <td>93</td>\n",
       "      <td>473</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>7.966172</td>\n",
       "      <td>11.674365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>TGATTTTT</td>\n",
       "      <td>521</td>\n",
       "      <td>2606</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>43.889733</td>\n",
       "      <td>11.870658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>GGTCT</td>\n",
       "      <td>571</td>\n",
       "      <td>2677</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>45.085501</td>\n",
       "      <td>12.664825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>ATTTCCTT</td>\n",
       "      <td>199</td>\n",
       "      <td>863</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>14.534474</td>\n",
       "      <td>13.691586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "12       AAAATGT      1249      7343       0.000567      123.669344   \n",
       "595       CTGTGG        89       512       0.000040        8.623002   \n",
       "124      TCAGTAT       410      2348       0.000181       39.544548   \n",
       "132       AGATGG       389      2227       0.000172       37.506691   \n",
       "14        TTGCAA      1203      6755       0.000521      113.766365   \n",
       "2         CAATTT      1955     10968       0.000847      184.720872   \n",
       "150      TTATATA       354      1983       0.000153       33.397291   \n",
       "783   AAGAAAAAAA        55       307       0.000024        5.170433   \n",
       "1030    TCCTTTCC        30       167       0.000013        2.812581   \n",
       "131       CTGACT       393      2104       0.000162       35.435149   \n",
       "574       GCTCAG        93       473       0.000037        7.966172   \n",
       "80      TGATTTTT       521      2606       0.000201       43.889733   \n",
       "63         GGTCT       571      2677       0.000207       45.085501   \n",
       "305     ATTTCCTT       199       863       0.000067       14.534474   \n",
       "\n",
       "      ratio_difference  \n",
       "12           10.099512  \n",
       "595          10.321231  \n",
       "124          10.368054  \n",
       "132          10.371483  \n",
       "14           10.574303  \n",
       "2            10.583536  \n",
       "150          10.599662  \n",
       "783          10.637407  \n",
       "1030         10.666360  \n",
       "131          11.090683  \n",
       "574          11.674365  \n",
       "80           11.870658  \n",
       "63           12.664825  \n",
       "305          13.691586  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 10]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_type</th>\n",
       "      <th>gene</th>\n",
       "      <th>name</th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>s35</td>\n",
       "      <td>Staphylococcus sRNA 35 (srn_0335)</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>CDS</td>\n",
       "      <td>lys2b</td>\n",
       "      <td>Thioester reductase domain of alpha aminoadipa...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>CDS</td>\n",
       "      <td>era</td>\n",
       "      <td>GTPase Era%2C involved in 16S rRNA processing</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961</th>\n",
       "      <td>CDS</td>\n",
       "      <td>rlha</td>\n",
       "      <td>23S rRNA C2501 and tRNA U34 5'-hydroxylation p...</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>CDS</td>\n",
       "      <td>dnag</td>\n",
       "      <td>DNA primase</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>CDS</td>\n",
       "      <td>degq</td>\n",
       "      <td>Serine protease HtrA-like</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CDS</td>\n",
       "      <td>adda</td>\n",
       "      <td>helicase-exonuclease AddAB subunit AddA</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>CDS</td>\n",
       "      <td>pepf</td>\n",
       "      <td>oligoendopeptidase F</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>CDS</td>\n",
       "      <td>vals</td>\n",
       "      <td>valine--tRNA ligase</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>CDS</td>\n",
       "      <td>nrde</td>\n",
       "      <td>class 1b ribonucleoside-diphosphate reductase ...</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>CDS</td>\n",
       "      <td>pola</td>\n",
       "      <td>DNA polymerase I</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>CDS</td>\n",
       "      <td>ftsk</td>\n",
       "      <td>DNA segregation ATPase FtsK/SpoIIIE or related...</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>979</th>\n",
       "      <td>CDS</td>\n",
       "      <td>rpir</td>\n",
       "      <td>DNA-binding transcriptional regulator%2C MurR/...</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>CDS</td>\n",
       "      <td>mvim</td>\n",
       "      <td>putative dehydrogenase</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>CDS</td>\n",
       "      <td>pyca</td>\n",
       "      <td>pyruvate carboxylase</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     feature_type   gene                                               name  \\\n",
       "1391        ncRNA    s35                  Staphylococcus sRNA 35 (srn_0335)   \n",
       "596           CDS  lys2b  Thioester reductase domain of alpha aminoadipa...   \n",
       "308           CDS    era      GTPase Era%2C involved in 16S rRNA processing   \n",
       "961           CDS   rlha  23S rRNA C2501 and tRNA U34 5'-hydroxylation p...   \n",
       "272           CDS   dnag                                        DNA primase   \n",
       "251           CDS   degq                          Serine protease HtrA-like   \n",
       "27            CDS   adda            helicase-exonuclease AddAB subunit AddA   \n",
       "810           CDS   pepf                               oligoendopeptidase F   \n",
       "1247          CDS   vals                                valine--tRNA ligase   \n",
       "758           CDS   nrde  class 1b ribonucleoside-diphosphate reductase ...   \n",
       "850           CDS   pola                                   DNA polymerase I   \n",
       "390           CDS   ftsk  DNA segregation ATPase FtsK/SpoIIIE or related...   \n",
       "979           CDS   rpir  DNA-binding transcriptional regulator%2C MurR/...   \n",
       "704           CDS   mvim                             putative dehydrogenase   \n",
       "906           CDS   pyca                               pyruvate carboxylase   \n",
       "\n",
       "      tokens  \n",
       "1391     289  \n",
       "596      229  \n",
       "308      220  \n",
       "961      210  \n",
       "272      192  \n",
       "251      182  \n",
       "27       181  \n",
       "810      173  \n",
       "1247     173  \n",
       "758      172  \n",
       "850      156  \n",
       "390      149  \n",
       "979      148  \n",
       "704      141  \n",
       "906      139  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tempDf = overrepresentedTokensDf.merge(overlappingDf[['tokens', 'feature_type', 'id', 'name', 'gene']], how='inner', on=['tokens'])\n",
    "\n",
    "tempDf[['tokens', 'feature_type', 'gene', 'name']].groupby(by=['feature_type', 'gene', 'name']).agg('count').reset_index().sort_values(by=['tokens'], ascending=False)[:15]\n",
    "# gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hyperosmolarity resistance protein Ebh</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Staphylococcus sRNA 35 (srn_0335)</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Thioester reductase domain of alpha aminoadipa...</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lipoprotein</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTPase Era%2C involved in 16S rRNA processing</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0             hyperosmolarity resistance protein Ebh    336\n",
       "1                  Staphylococcus sRNA 35 (srn_0335)    289\n",
       "2  Thioester reductase domain of alpha aminoadipa...    229\n",
       "3                                        Lipoprotein    221\n",
       "4      GTPase Era%2C involved in 16S rRNA processing    220"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/3615638238.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1642</th>\n",
       "      <td>ATGAT</td>\n",
       "      <td>8</td>\n",
       "      <td>67845</td>\n",
       "      <td>0.005237</td>\n",
       "      <td>1142.631981</td>\n",
       "      <td>0.007001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1747</th>\n",
       "      <td>ATCGC</td>\n",
       "      <td>6</td>\n",
       "      <td>43346</td>\n",
       "      <td>0.003346</td>\n",
       "      <td>730.024701</td>\n",
       "      <td>0.008219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529</th>\n",
       "      <td>ACAAT</td>\n",
       "      <td>10</td>\n",
       "      <td>49758</td>\n",
       "      <td>0.003841</td>\n",
       "      <td>838.014328</td>\n",
       "      <td>0.011933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>TACCT</td>\n",
       "      <td>5</td>\n",
       "      <td>17409</td>\n",
       "      <td>0.001344</td>\n",
       "      <td>293.198912</td>\n",
       "      <td>0.017053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1801</th>\n",
       "      <td>ACGTT</td>\n",
       "      <td>6</td>\n",
       "      <td>20409</td>\n",
       "      <td>0.001575</td>\n",
       "      <td>343.724314</td>\n",
       "      <td>0.017456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1230</th>\n",
       "      <td>AAGTAGCT</td>\n",
       "      <td>20</td>\n",
       "      <td>1202</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>20.243845</td>\n",
       "      <td>0.987955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>TTTTTCT</td>\n",
       "      <td>41</td>\n",
       "      <td>2461</td>\n",
       "      <td>0.000190</td>\n",
       "      <td>41.447672</td>\n",
       "      <td>0.989199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>AAGCTTA</td>\n",
       "      <td>44</td>\n",
       "      <td>2641</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>44.479196</td>\n",
       "      <td>0.989227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1319</th>\n",
       "      <td>AAGTCCT</td>\n",
       "      <td>16</td>\n",
       "      <td>953</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>16.050236</td>\n",
       "      <td>0.996870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>TAATTTT</td>\n",
       "      <td>266</td>\n",
       "      <td>15818</td>\n",
       "      <td>0.001221</td>\n",
       "      <td>266.403606</td>\n",
       "      <td>0.998485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1045 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "1642     ATGAT         8     67845       0.005237     1142.631981   \n",
       "1747     ATCGC         6     43346       0.003346      730.024701   \n",
       "1529     ACAAT        10     49758       0.003841      838.014328   \n",
       "1836     TACCT         5     17409       0.001344      293.198912   \n",
       "1801     ACGTT         6     20409       0.001575      343.724314   \n",
       "...        ...       ...       ...            ...             ...   \n",
       "1230  AAGTAGCT        20      1202       0.000093       20.243845   \n",
       "893    TTTTTCT        41      2461       0.000190       41.447672   \n",
       "861    AAGCTTA        44      2641       0.000204       44.479196   \n",
       "1319   AAGTCCT        16       953       0.000074       16.050236   \n",
       "217    TAATTTT       266     15818       0.001221      266.403606   \n",
       "\n",
       "      ratio_difference  \n",
       "1642          0.007001  \n",
       "1747          0.008219  \n",
       "1529          0.011933  \n",
       "1836          0.017053  \n",
       "1801          0.017456  \n",
       "...                ...  \n",
       "1230          0.987955  \n",
       "893           0.989199  \n",
       "861           0.989227  \n",
       "1319          0.996870  \n",
       "217           0.998485  \n",
       "\n",
       "[1045 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 1]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lys2b</td>\n",
       "      <td>28136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>araj</td>\n",
       "      <td>23981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pote</td>\n",
       "      <td>23001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ccma</td>\n",
       "      <td>22015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ebh</td>\n",
       "      <td>21359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  count\n",
       "0  lys2b  28136\n",
       "1   araj  23981\n",
       "2   pote  23001\n",
       "3   ccma  22015\n",
       "4    ebh  21359"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hyperosmolarity resistance protein Ebh</td>\n",
       "      <td>50615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Thioester reductase domain of alpha aminoadipa...</td>\n",
       "      <td>28136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lipoprotein</td>\n",
       "      <td>27845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Extracellular matrix-binding protein ebh</td>\n",
       "      <td>27736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>putative arabinose efflux permease AraJ%2C MFS...</td>\n",
       "      <td>24507</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0             hyperosmolarity resistance protein Ebh  50615\n",
       "1  Thioester reductase domain of alpha aminoadipa...  28136\n",
       "2                                        Lipoprotein  27845\n",
       "3           Extracellular matrix-binding protein ebh  27736\n",
       "4  putative arabinose efflux permease AraJ%2C MFS...  24507"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(-0.005417008772367425)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff = overlappingDf.score.mean() - 1 * overlappingDf.score.std()\n",
    "cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2222337.507815295 0.0 True\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/56670797.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 80]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>TAACAA</td>\n",
       "      <td>6859</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>84.532468</td>\n",
       "      <td>81.140420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>CATTG</td>\n",
       "      <td>6964</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>84.532468</td>\n",
       "      <td>82.382546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>GTCTG</td>\n",
       "      <td>11695</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>135.251950</td>\n",
       "      <td>86.468255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>ATATTTG</td>\n",
       "      <td>7640</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>84.532468</td>\n",
       "      <td>90.379474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>AATTAAA</td>\n",
       "      <td>12563</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>135.251950</td>\n",
       "      <td>92.885907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CACCAC</td>\n",
       "      <td>12850</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>135.251950</td>\n",
       "      <td>95.007873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ACAGT</td>\n",
       "      <td>12057</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>118.345456</td>\n",
       "      <td>101.879704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>AATTT</td>\n",
       "      <td>15299</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>118.345456</td>\n",
       "      <td>129.274081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TGATTT</td>\n",
       "      <td>31651</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>202.877924</td>\n",
       "      <td>156.010074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>TAATTT</td>\n",
       "      <td>19852</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>118.345456</td>\n",
       "      <td>167.746196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "499   TAACAA      6859         5       0.000006       84.532468   \n",
       "486    CATTG      6964         5       0.000006       84.532468   \n",
       "181    GTCTG     11695         8       0.000010      135.251950   \n",
       "417  ATATTTG      7640         5       0.000006       84.532468   \n",
       "140  AATTAAA     12563         8       0.000010      135.251950   \n",
       "131   CACCAC     12850         8       0.000010      135.251950   \n",
       "167    ACAGT     12057         7       0.000009      118.345456   \n",
       "78     AATTT     15299         7       0.000009      118.345456   \n",
       "13    TGATTT     31651        12       0.000015      202.877924   \n",
       "49    TAATTT     19852         7       0.000009      118.345456   \n",
       "\n",
       "     ratio_difference  \n",
       "499         81.140420  \n",
       "486         82.382546  \n",
       "181         86.468255  \n",
       "417         90.379474  \n",
       "140         92.885907  \n",
       "131         95.007873  \n",
       "167        101.879704  \n",
       "78         129.274081  \n",
       "13         156.010074  \n",
       "49         167.746196  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 80]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lys2b</td>\n",
       "      <td>727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>araj</td>\n",
       "      <td>559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccma</td>\n",
       "      <td>554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>caia</td>\n",
       "      <td>531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ftsk</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  count\n",
       "0  lys2b    727\n",
       "1   araj    559\n",
       "2   ccma    554\n",
       "3   caia    531\n",
       "4   ftsk    482"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hyperosmolarity resistance protein Ebh</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Thioester reductase domain of alpha aminoadipa...</td>\n",
       "      <td>727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Extracellular matrix-binding protein ebh</td>\n",
       "      <td>612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lipoprotein</td>\n",
       "      <td>605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>putative arabinose efflux permease AraJ%2C MFS...</td>\n",
       "      <td>573</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0             hyperosmolarity resistance protein Ebh    967\n",
       "1  Thioester reductase domain of alpha aminoadipa...    727\n",
       "2           Extracellular matrix-binding protein ebh    612\n",
       "3                                        Lipoprotein    605\n",
       "4  putative arabinose efflux permease AraJ%2C MFS...    573"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21404/3155434324.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.3]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3102</th>\n",
       "      <td>GCTCAG</td>\n",
       "      <td>465</td>\n",
       "      <td>101</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>1707.555863</td>\n",
       "      <td>0.272319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3080</th>\n",
       "      <td>CTGTGG</td>\n",
       "      <td>496</td>\n",
       "      <td>105</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>1775.181837</td>\n",
       "      <td>0.279408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2737</th>\n",
       "      <td>TTTCTTTTTT</td>\n",
       "      <td>1030</td>\n",
       "      <td>218</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>3685.615624</td>\n",
       "      <td>0.279465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>GGAGATTTCA</td>\n",
       "      <td>57</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>202.877924</td>\n",
       "      <td>0.280957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3261</th>\n",
       "      <td>GTGATCTG</td>\n",
       "      <td>227</td>\n",
       "      <td>47</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>794.605203</td>\n",
       "      <td>0.285676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "3102      GCTCAG       465       101       0.000125     1707.555863   \n",
       "3080      CTGTGG       496       105       0.000130     1775.181837   \n",
       "2737  TTTCTTTTTT      1030       218       0.000271     3685.615624   \n",
       "3370  GGAGATTTCA        57        12       0.000015      202.877924   \n",
       "3261    GTGATCTG       227        47       0.000058      794.605203   \n",
       "\n",
       "      ratio_difference  \n",
       "3102          0.272319  \n",
       "3080          0.279408  \n",
       "2737          0.279465  \n",
       "3370          0.280957  \n",
       "3261          0.285676  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.3]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mnth</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dnak</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ybjt</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tagb</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>amya</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0  mnth     65\n",
       "1  dnak     53\n",
       "2  ybjt     43\n",
       "3  tagb     35\n",
       "4  amya     33"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lipoprotein</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>molecular chaperone DnaK</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mn2+ or Fe2+ transporter%2C NRAMP family</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DUF2867 domain-containing protein</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDP-glycerol glycerophosphotransferase%2C TagB...</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0                                        Lipoprotein     72\n",
       "1                           molecular chaperone DnaK     53\n",
       "2           Mn2+ or Fe2+ transporter%2C NRAMP family     52\n",
       "3                  DUF2867 domain-containing protein     43\n",
       "4  CDP-glycerol glycerophosphotransferase%2C TagB...     35"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
