{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistical analysis to get the significantly associated tokens for attribution score groupings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read mapping information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>visit_occurrence_id</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>COUNTRY_OF_BIRTH</th>\n",
       "      <th>SEX</th>\n",
       "      <th>Age_Bin</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>678487</td>\n",
       "      <td>3312694</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>3312694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2094045</td>\n",
       "      <td>9199249</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(20, 30]</td>\n",
       "      <td>9199249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2107813</td>\n",
       "      <td>9410987</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(50, 60]</td>\n",
       "      <td>9410987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2083212</td>\n",
       "      <td>602890</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2083212</td>\n",
       "      <td>9039437</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(40, 50]</td>\n",
       "      <td>9039437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7194</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16588204</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16588204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7195</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16709341</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16709341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7196</th>\n",
       "      <td>2603535</td>\n",
       "      <td>16651755</td>\n",
       "      <td>Inpatient</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Male</td>\n",
       "      <td>(60, 70]</td>\n",
       "      <td>16651755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7197</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16872774</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16872774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198</th>\n",
       "      <td>2098616</td>\n",
       "      <td>16869916</td>\n",
       "      <td>Emergency</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Female</td>\n",
       "      <td>(80, 90]</td>\n",
       "      <td>16869916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7199 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      person_id  visit_occurrence_id       TYPE COUNTRY_OF_BIRTH     SEX  \\\n",
       "0        678487              3312694  Emergency        Australia    Male   \n",
       "1       2094045              9199249  Emergency        Australia    Male   \n",
       "2       2107813              9410987  Emergency        Australia    Male   \n",
       "3       2083212               602890  Inpatient        Australia  Female   \n",
       "4       2083212              9039437  Emergency        Australia  Female   \n",
       "...         ...                  ...        ...              ...     ...   \n",
       "7194    2603535             16588204  Inpatient        Australia    Male   \n",
       "7195    2603535             16709341  Inpatient        Australia    Male   \n",
       "7196    2603535             16651755  Inpatient        Australia    Male   \n",
       "7197    2098616             16872774  Emergency        Australia  Female   \n",
       "7198    2098616             16869916  Emergency        Australia  Female   \n",
       "\n",
       "       Age_Bin  JOURNEY_ID  \n",
       "0     (60, 70]     3312694  \n",
       "1     (20, 30]     9199249  \n",
       "2     (50, 60]     9410987  \n",
       "3     (40, 50]     9039437  \n",
       "4     (40, 50]     9039437  \n",
       "...        ...         ...  \n",
       "7194  (60, 70]    16588204  \n",
       "7195  (60, 70]    16709341  \n",
       "7196  (60, 70]    16651755  \n",
       "7197  (80, 90]    16872774  \n",
       "7198  (80, 90]    16869916  \n",
       "\n",
       "[7199 rows x 7 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "demoTrainDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_train.csv', sep='\\t')\n",
    "demoTestDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_test.csv', sep='\\t')\n",
    "demoValidateDf = pd.read_csv(os.environ['EHR_DATA_BASE'] + '/blood_pos_cohort_20240614/data/wb_365_wa_1/splits_v1/demographics/mortality_normal_validate.csv', sep='\\t')\n",
    "demoDf = pd.concat([demoTrainDf, demoTestDf, demoValidateDf], ignore_index=True)\n",
    "demoDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tube_code</th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>db_ID</th>\n",
       "      <th>pt_age</th>\n",
       "      <th>pt_gender</th>\n",
       "      <th>EPISODE_ID</th>\n",
       "      <th>hospital_admission</th>\n",
       "      <th>hospital_discharge</th>\n",
       "      <th>hospital_in_last_year</th>\n",
       "      <th>genome_species</th>\n",
       "      <th>species_reported</th>\n",
       "      <th>contig_number</th>\n",
       "      <th>length</th>\n",
       "      <th>tube_code_duplicate_or_old</th>\n",
       "      <th>date_of_collection</th>\n",
       "      <th>ID_number</th>\n",
       "      <th>location_additional_02</th>\n",
       "      <th>collected_from_original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AH19J072</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>92</td>\n",
       "      <td>5131021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-0997</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AH19J074</td>\n",
       "      <td>213972.0</td>\n",
       "      <td>3JNQS</td>\n",
       "      <td>64</td>\n",
       "      <td>Male</td>\n",
       "      <td>12931474.0</td>\n",
       "      <td>2019-10-17</td>\n",
       "      <td>2019-11-29</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus sp. D19</td>\n",
       "      <td>Streptococcus mitis group</td>\n",
       "      <td>84</td>\n",
       "      <td>2022991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-10-30</td>\n",
       "      <td>19-303-1174</td>\n",
       "      <td>A-7EA;HAEM</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AH19B003</td>\n",
       "      <td>526238.0</td>\n",
       "      <td>567AE</td>\n",
       "      <td>85</td>\n",
       "      <td>Male</td>\n",
       "      <td>640261.0</td>\n",
       "      <td>2019-02-04</td>\n",
       "      <td>2019-02-07</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>93</td>\n",
       "      <td>4942915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-02-03</td>\n",
       "      <td>19-034-1596</td>\n",
       "      <td>A-4WB;GMC</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH20A024</td>\n",
       "      <td>788941.0</td>\n",
       "      <td>3YBDS</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>13194824.0</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>no</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>110</td>\n",
       "      <td>5149157</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-16</td>\n",
       "      <td>20-016-2112</td>\n",
       "      <td>S-ED;Emergency</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH20L041</td>\n",
       "      <td>1023735.0</td>\n",
       "      <td>2VJ78</td>\n",
       "      <td>42</td>\n",
       "      <td>Male</td>\n",
       "      <td>14210147.0</td>\n",
       "      <td>2020-12-18</td>\n",
       "      <td>2021-01-05</td>\n",
       "      <td>yes</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>46</td>\n",
       "      <td>2735216</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-16</td>\n",
       "      <td>20-351-2856</td>\n",
       "      <td>Emergency Dept;nan</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3112</th>\n",
       "      <td>ALF22L076</td>\n",
       "      <td>639020.0</td>\n",
       "      <td>3S6HL</td>\n",
       "      <td>84</td>\n",
       "      <td>Male</td>\n",
       "      <td>17219899.0</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>2022-12-20</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus vestibularis</td>\n",
       "      <td>Streptococcus salivarius group</td>\n",
       "      <td>107</td>\n",
       "      <td>1938627</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-15</td>\n",
       "      <td>22-349-2150</td>\n",
       "      <td>Ward 4WB</td>\n",
       "      <td>Blood Peripheral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>ALF22L089</td>\n",
       "      <td>655404.0</td>\n",
       "      <td>6EWBM</td>\n",
       "      <td>67</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus columbae</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>152</td>\n",
       "      <td>3972219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-17</td>\n",
       "      <td>22-351-1465</td>\n",
       "      <td>Comm/Amb Clinic</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3114</th>\n",
       "      <td>ALF22L138</td>\n",
       "      <td>2572048.0</td>\n",
       "      <td>YM5PJ</td>\n",
       "      <td>61</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter cloacae complex</td>\n",
       "      <td>66</td>\n",
       "      <td>4876179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-12-27</td>\n",
       "      <td>22-361-0202</td>\n",
       "      <td>Ward 7 East</td>\n",
       "      <td>Hick White Lum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3115</th>\n",
       "      <td>ALF23A102</td>\n",
       "      <td>2054964.0</td>\n",
       "      <td>5DYAH</td>\n",
       "      <td>35</td>\n",
       "      <td>Female</td>\n",
       "      <td>17294287.0</td>\n",
       "      <td>2023-01-12</td>\n",
       "      <td>2023-02-01</td>\n",
       "      <td>yes</td>\n",
       "      <td>Streptococcus constellatus</td>\n",
       "      <td>Streptococcus anginosus group</td>\n",
       "      <td>54</td>\n",
       "      <td>1869828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-11</td>\n",
       "      <td>23-011-3305</td>\n",
       "      <td>ICU</td>\n",
       "      <td>Blood Venous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>ALF23B102</td>\n",
       "      <td>2015952.0</td>\n",
       "      <td>57E2W</td>\n",
       "      <td>82</td>\n",
       "      <td>Female</td>\n",
       "      <td>17281239.0</td>\n",
       "      <td>2023-01-08</td>\n",
       "      <td>2023-02-16</td>\n",
       "      <td>no</td>\n",
       "      <td>Acinetobacter colistiniresistens</td>\n",
       "      <td>Acinetobacter species</td>\n",
       "      <td>233</td>\n",
       "      <td>3952992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-15</td>\n",
       "      <td>23-046-3310</td>\n",
       "      <td>ICU</td>\n",
       "      <td>PICC Line White</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3117 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      tube_code  PATIENT_ID  db_ID  pt_age pt_gender  EPISODE_ID  \\\n",
       "0      AH19J072    213972.0  3JNQS      64      Male  12931474.0   \n",
       "1      AH19J074    213972.0  3JNQS      64      Male  12931474.0   \n",
       "2      AH19B003    526238.0  567AE      85      Male    640261.0   \n",
       "3      AH20A024    788941.0  3YBDS      58    Female  13194824.0   \n",
       "4      AH20L041   1023735.0  2VJ78      42      Male  14210147.0   \n",
       "...         ...         ...    ...     ...       ...         ...   \n",
       "3112  ALF22L076    639020.0  3S6HL      84      Male  17219899.0   \n",
       "3113  ALF22L089    655404.0  6EWBM      67    Female         NaN   \n",
       "3114  ALF22L138   2572048.0  YM5PJ      61      Male         NaN   \n",
       "3115  ALF23A102   2054964.0  5DYAH      35    Female  17294287.0   \n",
       "3116  ALF23B102   2015952.0  57E2W      82    Female  17281239.0   \n",
       "\n",
       "     hospital_admission hospital_discharge hospital_in_last_year  \\\n",
       "0            2019-10-17         2019-11-29                   yes   \n",
       "1            2019-10-17         2019-11-29                   yes   \n",
       "2            2019-02-04         2019-02-07                    no   \n",
       "3            2020-01-16         2020-01-19                    no   \n",
       "4            2020-12-18         2021-01-05                   yes   \n",
       "...                 ...                ...                   ...   \n",
       "3112         2022-12-15         2022-12-20                   yes   \n",
       "3113                NaN                NaN                   NaN   \n",
       "3114                NaN                NaN                   NaN   \n",
       "3115         2023-01-12         2023-02-01                   yes   \n",
       "3116         2023-01-08         2023-02-16                    no   \n",
       "\n",
       "                        genome_species                species_reported  \\\n",
       "0                     Escherichia coli                Escherichia coli   \n",
       "1                Streptococcus sp. D19       Streptococcus mitis group   \n",
       "2                     Escherichia coli                Escherichia coli   \n",
       "3                     Escherichia coli                Escherichia coli   \n",
       "4                Staphylococcus aureus           Staphylococcus aureus   \n",
       "...                                ...                             ...   \n",
       "3112        Streptococcus vestibularis  Streptococcus salivarius group   \n",
       "3113                  Proteus columbae                Proteus vulgaris   \n",
       "3114         Enterobacter roggenkampii    Enterobacter cloacae complex   \n",
       "3115        Streptococcus constellatus   Streptococcus anginosus group   \n",
       "3116  Acinetobacter colistiniresistens           Acinetobacter species   \n",
       "\n",
       "      contig_number   length tube_code_duplicate_or_old date_of_collection  \\\n",
       "0                92  5131021                        NaN         2019-10-30   \n",
       "1                84  2022991                        NaN         2019-10-30   \n",
       "2                93  4942915                        NaN         2019-02-03   \n",
       "3               110  5149157                        NaN         2020-01-16   \n",
       "4                46  2735216                        NaN         2020-12-16   \n",
       "...             ...      ...                        ...                ...   \n",
       "3112            107  1938627                        NaN         2022-12-15   \n",
       "3113            152  3972219                        NaN         2022-12-17   \n",
       "3114             66  4876179                        NaN         2022-12-27   \n",
       "3115             54  1869828                        NaN         2023-01-11   \n",
       "3116            233  3952992                        NaN         2023-02-15   \n",
       "\n",
       "        ID_number location_additional_02 collected_from_original  \n",
       "0     19-303-0997             A-7EA;HAEM                   Blood  \n",
       "1     19-303-1174             A-7EA;HAEM        Blood Peripheral  \n",
       "2     19-034-1596              A-4WB;GMC                   Blood  \n",
       "3     20-016-2112         S-ED;Emergency            Blood Venous  \n",
       "4     20-351-2856     Emergency Dept;nan                   Blood  \n",
       "...           ...                    ...                     ...  \n",
       "3112  22-349-2150               Ward 4WB        Blood Peripheral  \n",
       "3113  22-351-1465        Comm/Amb Clinic                   Blood  \n",
       "3114  22-361-0202            Ward 7 East          Hick White Lum  \n",
       "3115  23-011-3305                    ICU            Blood Venous  \n",
       "3116  23-046-3310                    ICU         PICC Line White  \n",
       "\n",
       "[3117 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "mappingDf = pd.read_csv(os.environ['GENOMICS_DATA_BASE'] + '/patient_tube_id_mapping_full.tsv', sep='\\t')\n",
    "mappingDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "      <th>tube_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50056</td>\n",
       "      <td>13224842</td>\n",
       "      <td>AH20B011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62567</td>\n",
       "      <td>12645629</td>\n",
       "      <td>AH19G065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>138108</td>\n",
       "      <td>14737781</td>\n",
       "      <td>AH21E052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>AH20J009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>211314</td>\n",
       "      <td>13955099</td>\n",
       "      <td>KPN2214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>2084373</td>\n",
       "      <td>16158768</td>\n",
       "      <td>ALF22C004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>1058066</td>\n",
       "      <td>16759482</td>\n",
       "      <td>ALF22H089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>2552781</td>\n",
       "      <td>16966609</td>\n",
       "      <td>ALF22J054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2638484</td>\n",
       "      <td>16969999</td>\n",
       "      <td>ALF22K143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>2015952</td>\n",
       "      <td>17281239</td>\n",
       "      <td>ALF23B102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>517 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     person_id  JOURNEY_ID  tube_code\n",
       "0        50056    13224842   AH20B011\n",
       "1        62567    12645629   AH19G065\n",
       "2       138108    14737781   AH21E052\n",
       "3       211314    13955099   AH20J009\n",
       "4       211314    13955099    KPN2214\n",
       "..         ...         ...        ...\n",
       "512    2084373    16158768  ALF22C004\n",
       "513    1058066    16759482  ALF22H089\n",
       "514    2552781    16966609  ALF22J054\n",
       "515    2638484    16969999  ALF22K143\n",
       "516    2015952    17281239  ALF23B102\n",
       "\n",
       "[517 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mappedJourneyTubeIdsDf = mappingDf[['tube_code', 'PATIENT_ID', 'EPISODE_ID']].drop_duplicates().merge(\n",
    "    demoDf[['person_id', 'visit_occurrence_id', 'JOURNEY_ID']],\n",
    "    how='inner',\n",
    "    left_on=['PATIENT_ID', 'EPISODE_ID'],\n",
    "    right_on=['person_id', 'visit_occurrence_id']\n",
    ").drop(\n",
    "    columns=['PATIENT_ID', 'EPISODE_ID', 'visit_occurrence_id']\n",
    ")[['person_id', 'JOURNEY_ID', 'tube_code']].drop_duplicates()\n",
    "mappedJourneyTubeIdsDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read bed files containing overlapping tokens and attribution scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>tokens</th>\n",
       "      <th>score</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>atributes</th>\n",
       "      <th>tube_code</th>\n",
       "      <th>person_id</th>\n",
       "      <th>JOURNEY_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>191</td>\n",
       "      <td>192</td>\n",
       "      <td>GT</td>\n",
       "      <td>-0.001426</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>193</td>\n",
       "      <td>197</td>\n",
       "      <td>ATAAC</td>\n",
       "      <td>-0.005218</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>198</td>\n",
       "      <td>199</td>\n",
       "      <td>GC</td>\n",
       "      <td>-0.000682</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>203</td>\n",
       "      <td>GCCA</td>\n",
       "      <td>-0.010548</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>204</td>\n",
       "      <td>210</td>\n",
       "      <td>ATTGATT</td>\n",
       "      <td>-0.004120</td>\n",
       "      <td>CFBAKF_00005</td>\n",
       "      <td>DUF3969 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ID=CFBAKF_00005;Name=DUF3969 domain-containing...</td>\n",
       "      <td>AH19E065</td>\n",
       "      <td>2221447</td>\n",
       "      <td>681015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121252</th>\n",
       "      <td>31</td>\n",
       "      <td>99</td>\n",
       "      <td>103</td>\n",
       "      <td>GGTGG</td>\n",
       "      <td>-0.252315</td>\n",
       "      <td>KKIOAG_13715</td>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>trnI</td>\n",
       "      <td>ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121253</th>\n",
       "      <td>31</td>\n",
       "      <td>104</td>\n",
       "      <td>106</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-0.134230</td>\n",
       "      <td>KKIOAG_13715</td>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>trnI</td>\n",
       "      <td>ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121254</th>\n",
       "      <td>31</td>\n",
       "      <td>107</td>\n",
       "      <td>110</td>\n",
       "      <td>GAGT</td>\n",
       "      <td>-0.080241</td>\n",
       "      <td>KKIOAG_13715</td>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>trnI</td>\n",
       "      <td>ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121255</th>\n",
       "      <td>31</td>\n",
       "      <td>111</td>\n",
       "      <td>116</td>\n",
       "      <td>CCACTT</td>\n",
       "      <td>0.215479</td>\n",
       "      <td>KKIOAG_13715</td>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>trnI</td>\n",
       "      <td>ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121256</th>\n",
       "      <td>31</td>\n",
       "      <td>117</td>\n",
       "      <td>120</td>\n",
       "      <td>AGGC</td>\n",
       "      <td>-0.087724</td>\n",
       "      <td>KKIOAG_13715</td>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>trnI</td>\n",
       "      <td>ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...</td>\n",
       "      <td>AH21K020</td>\n",
       "      <td>2035444</td>\n",
       "      <td>15586881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121257 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig_id  start_position  end_position   tokens     score  \\\n",
       "0               1             191           192       GT -0.001426   \n",
       "1               1             193           197    ATAAC -0.005218   \n",
       "2               1             198           199       GC -0.000682   \n",
       "3               1             200           203     GCCA -0.010548   \n",
       "4               1             204           210  ATTGATT -0.004120   \n",
       "...           ...             ...           ...      ...       ...   \n",
       "121252         31              99           103    GGTGG -0.252315   \n",
       "121253         31             104           106      TTC -0.134230   \n",
       "121254         31             107           110     GAGT -0.080241   \n",
       "121255         31             111           116   CCACTT  0.215479   \n",
       "121256         31             117           120     AGGC -0.087724   \n",
       "\n",
       "                  id                               name  gene  \\\n",
       "0       CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "1       CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "2       CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "3       CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "4       CFBAKF_00005  DUF3969 domain-containing protein   NaN   \n",
       "...              ...                                ...   ...   \n",
       "121252  KKIOAG_13715                      tRNA-Ile(gat)  trnI   \n",
       "121253  KKIOAG_13715                      tRNA-Ile(gat)  trnI   \n",
       "121254  KKIOAG_13715                      tRNA-Ile(gat)  trnI   \n",
       "121255  KKIOAG_13715                      tRNA-Ile(gat)  trnI   \n",
       "121256  KKIOAG_13715                      tRNA-Ile(gat)  trnI   \n",
       "\n",
       "                                                atributes tube_code  \\\n",
       "0       ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065   \n",
       "1       ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065   \n",
       "2       ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065   \n",
       "3       ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065   \n",
       "4       ID=CFBAKF_00005;Name=DUF3969 domain-containing...  AH19E065   \n",
       "...                                                   ...       ...   \n",
       "121252  ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...  AH21K020   \n",
       "121253  ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...  AH21K020   \n",
       "121254  ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...  AH21K020   \n",
       "121255  ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...  AH21K020   \n",
       "121256  ID=KKIOAG_13715;Name=tRNA-Ile(gat);locus_tag=K...  AH21K020   \n",
       "\n",
       "        person_id  JOURNEY_ID  \n",
       "0         2221447      681015  \n",
       "1         2221447      681015  \n",
       "2         2221447      681015  \n",
       "3         2221447      681015  \n",
       "4         2221447      681015  \n",
       "...           ...         ...  \n",
       "121252    2035444    15586881  \n",
       "121253    2035444    15586881  \n",
       "121254    2035444    15586881  \n",
       "121255    2035444    15586881  \n",
       "121256    2035444    15586881  \n",
       "\n",
       "[121257 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "overlappingFilesDir = Path(os.environ['GENOMICS_DATA_BASE'], 'genome_nlp_tokens', 'overlapping_with_annotations')\n",
    "\n",
    "overlappingDfList = []\n",
    "for overlappingFile in os.listdir(overlappingFilesDir):\n",
    "    df = pd.read_csv(Path(overlappingFilesDir, overlappingFile), sep='\\t', names=['contig_id', 'start_position', 'end_position', 'tokens', 'score', 'id', 'name', 'gene', 'atributes'])\n",
    "    df['tube_code'] = [overlappingFile.split('_')[0]]*df.shape[0]\n",
    "    overlappingDfList.append(df)\n",
    "overlappingDf = pd.concat(overlappingDfList, ignore_index=True)\n",
    "overlappingDf = overlappingDf.merge(\n",
    "    mappedJourneyTubeIdsDf,\n",
    "    how='inner',\n",
    "    on=['tube_code']\n",
    ")\n",
    "overlappingDf = overlappingDf.drop_duplicates()\n",
    "overlappingDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform chi-square test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1197.7722222222224 0.020681173280632906 True\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "cutoff = 0.02\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`The test indicates that the tokens associated with the high score group are significantly different than the rest`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/2494458605.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>AAAATGT</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>1.241343</td>\n",
       "      <td>4.027895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CCTGAG</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>0.000833</td>\n",
       "      <td>1.179276</td>\n",
       "      <td>4.239890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>TTGGTT</td>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>0.000964</td>\n",
       "      <td>1.365477</td>\n",
       "      <td>4.394068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TATAGT</td>\n",
       "      <td>9</td>\n",
       "      <td>31</td>\n",
       "      <td>0.001359</td>\n",
       "      <td>1.924082</td>\n",
       "      <td>4.677556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TAATTT</td>\n",
       "      <td>29</td>\n",
       "      <td>97</td>\n",
       "      <td>0.004252</td>\n",
       "      <td>6.020514</td>\n",
       "      <td>4.816865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CAC</td>\n",
       "      <td>27</td>\n",
       "      <td>89</td>\n",
       "      <td>0.003901</td>\n",
       "      <td>5.523977</td>\n",
       "      <td>4.887783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CTAA</td>\n",
       "      <td>12</td>\n",
       "      <td>37</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>2.296485</td>\n",
       "      <td>5.225378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ATGTGCA</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>1.117209</td>\n",
       "      <td>5.370527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CTGACT</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000789</td>\n",
       "      <td>1.117209</td>\n",
       "      <td>5.370527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGAGC</td>\n",
       "      <td>31</td>\n",
       "      <td>88</td>\n",
       "      <td>0.003857</td>\n",
       "      <td>5.461909</td>\n",
       "      <td>5.675671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>GAGCA</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>0.000657</td>\n",
       "      <td>0.931007</td>\n",
       "      <td>6.444633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TGCATT</td>\n",
       "      <td>13</td>\n",
       "      <td>32</td>\n",
       "      <td>0.001403</td>\n",
       "      <td>1.986149</td>\n",
       "      <td>6.545330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ATCC</td>\n",
       "      <td>26</td>\n",
       "      <td>59</td>\n",
       "      <td>0.002586</td>\n",
       "      <td>3.661962</td>\n",
       "      <td>7.100019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TTGA</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.682739</td>\n",
       "      <td>7.323446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CTCAGC</td>\n",
       "      <td>26</td>\n",
       "      <td>49</td>\n",
       "      <td>0.002148</td>\n",
       "      <td>3.041290</td>\n",
       "      <td>8.549003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TAACCA</td>\n",
       "      <td>24</td>\n",
       "      <td>43</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>2.668888</td>\n",
       "      <td>8.992511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GTGC</td>\n",
       "      <td>26</td>\n",
       "      <td>42</td>\n",
       "      <td>0.001841</td>\n",
       "      <td>2.606820</td>\n",
       "      <td>9.973836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CACTAT</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.682739</td>\n",
       "      <td>11.717514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AGATATTT</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.682739</td>\n",
       "      <td>11.717514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCGG</td>\n",
       "      <td>29</td>\n",
       "      <td>38</td>\n",
       "      <td>0.001666</td>\n",
       "      <td>2.358552</td>\n",
       "      <td>12.295681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ATGAAA</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>0.000701</td>\n",
       "      <td>0.993074</td>\n",
       "      <td>13.090660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ATTGAG</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>0.001140</td>\n",
       "      <td>1.613746</td>\n",
       "      <td>14.252553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ATCT</td>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>0.001052</td>\n",
       "      <td>1.489612</td>\n",
       "      <td>15.440266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>TGATCA</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.372403</td>\n",
       "      <td>16.111582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ATTCACAG</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.682739</td>\n",
       "      <td>16.111582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ACAAGG</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.372403</td>\n",
       "      <td>24.167373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TTCTTTGT</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000219</td>\n",
       "      <td>0.310336</td>\n",
       "      <td>29.000847</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "54   AAAATGT         5        20       0.000877        1.241343   \n",
       "57    CCTGAG         5        19       0.000833        1.179276   \n",
       "45    TTGGTT         6        22       0.000964        1.365477   \n",
       "25    TATAGT         9        31       0.001359        1.924082   \n",
       "4     TAATTT        29        97       0.004252        6.020514   \n",
       "5        CAC        27        89       0.003901        5.523977   \n",
       "22      CTAA        12        37       0.001622        2.296485   \n",
       "44   ATGTGCA         6        18       0.000789        1.117209   \n",
       "50    CTGACT         6        18       0.000789        1.117209   \n",
       "2      AGAGC        31        88       0.003857        5.461909   \n",
       "52     GAGCA         6        15       0.000657        0.931007   \n",
       "20    TGCATT        13        32       0.001403        1.986149   \n",
       "6       ATCC        26        59       0.002586        3.661962   \n",
       "64      TTGA         5        11       0.000482        0.682739   \n",
       "7     CTCAGC        26        49       0.002148        3.041290   \n",
       "10    TAACCA        24        43       0.001885        2.668888   \n",
       "8       GTGC        26        42       0.001841        2.606820   \n",
       "34    CACTAT         8        11       0.000482        0.682739   \n",
       "33  AGATATTT         8        11       0.000482        0.682739   \n",
       "3       CCGG        29        38       0.001666        2.358552   \n",
       "19    ATGAAA        13        16       0.000701        0.993074   \n",
       "12    ATTGAG        23        26       0.001140        1.613746   \n",
       "11      ATCT        23        24       0.001052        1.489612   \n",
       "51    TGATCA         6         6       0.000263        0.372403   \n",
       "23  ATTCACAG        11        11       0.000482        0.682739   \n",
       "27    ACAAGG         9         6       0.000263        0.372403   \n",
       "26  TTCTTTGT         9         5       0.000219        0.310336   \n",
       "\n",
       "    ratio_difference  \n",
       "54          4.027895  \n",
       "57          4.239890  \n",
       "45          4.394068  \n",
       "25          4.677556  \n",
       "4           4.816865  \n",
       "5           4.887783  \n",
       "22          5.225378  \n",
       "44          5.370527  \n",
       "50          5.370527  \n",
       "2           5.675671  \n",
       "52          6.444633  \n",
       "20          6.545330  \n",
       "6           7.100019  \n",
       "64          7.323446  \n",
       "7           8.549003  \n",
       "10          8.992511  \n",
       "8           9.973836  \n",
       "34         11.717514  \n",
       "33         11.717514  \n",
       "3          12.295681  \n",
       "19         13.090660  \n",
       "12         14.252553  \n",
       "11         15.440266  \n",
       "51         16.111582  \n",
       "23         16.111582  \n",
       "27         24.167373  \n",
       "26         29.000847  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trnI</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsr24</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trnA</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trnR</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trnL</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  count\n",
       "0   trnI    133\n",
       "1  tsr24     95\n",
       "2   trnA     84\n",
       "3   trnR     84\n",
       "4   trnL     77"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tRNA-Ile(gat)</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tRNA-Leu(taa)</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S. aureus tsr24 small RNA</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name  count\n",
       "0              tRNA-Ile(gat)    133\n",
       "1              tRNA-Ala(tgc)     84\n",
       "2              tRNA-Arg(acg)     84\n",
       "3              tRNA-Leu(taa)     76\n",
       "4  S. aureus tsr24 small RNA     56"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/3382577452.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.5]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>CGAA</td>\n",
       "      <td>5</td>\n",
       "      <td>241</td>\n",
       "      <td>0.010564</td>\n",
       "      <td>14.958184</td>\n",
       "      <td>0.334265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CG</td>\n",
       "      <td>25</td>\n",
       "      <td>1112</td>\n",
       "      <td>0.048742</td>\n",
       "      <td>69.018673</td>\n",
       "      <td>0.362221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tokens  count_hs  count_ls  proportion_ls  count_expected  ratio_difference\n",
       "63   CGAA         5       241       0.010564       14.958184          0.334265\n",
       "9      CG        25      1112       0.048742       69.018673          0.362221"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.5]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rrl</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rrs</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tnp</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trnA</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trnL</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0   rrl    143\n",
       "1   rrs    101\n",
       "2   tnp     90\n",
       "3  trnA     83\n",
       "4  trnL     38"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the high attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phage protein</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISL3 family IS1181 transposase</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             name  count\n",
       "0               23S ribosomal RNA     86\n",
       "1                   tRNA-Ala(tgc)     83\n",
       "2               16S ribosomal RNA     72\n",
       "3                   Phage protein     46\n",
       "4  ISL3 family IS1181 transposase     44"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11124.0714839337 6.47032501278991e-161 True\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "\n",
    "cutoff = -0.02\n",
    "highScoreOverlappingDf = overlappingDf[(overlappingDf.score > cutoff)].tokens.value_counts().reset_index()\n",
    "lowScoreOverlappingDf = overlappingDf[(overlappingDf.score < cutoff)].tokens.value_counts().reset_index()\n",
    "mergedOverlappingDf = highScoreOverlappingDf.add_suffix('_hs').merge(\n",
    "    lowScoreOverlappingDf.add_suffix('_ls'),\n",
    "    how='inner',\n",
    "    left_on=['tokens_hs'],\n",
    "    right_on=['tokens_ls']\n",
    ")[['tokens_hs', 'count_hs', 'count_ls']].rename(columns={'tokens_hs': 'tokens'})\n",
    "\n",
    "mergedOverlappingDf['proportion_ls'] = mergedOverlappingDf.count_ls/mergedOverlappingDf.count_ls.sum()\n",
    "\n",
    "mergedOverlappingDf['count_expected'] = mergedOverlappingDf.proportion_ls * mergedOverlappingDf.count_hs.sum()\n",
    "\n",
    "filteredOverlappingDf = mergedOverlappingDf[(mergedOverlappingDf.count_hs >= 5) & (mergedOverlappingDf.count_ls >= 5)]\n",
    "\n",
    "chi2, p, dof, expected = stats.chi2_contingency(pd.crosstab(filteredOverlappingDf.count_hs, filteredOverlappingDf.count_expected), correction=True)\n",
    "significant = p < 0.05  # 5% significance level\n",
    "print(chi2, p, significant)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate the ratio of count differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/3361415802.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)\n"
     ]
    }
   ],
   "source": [
    "filteredOverlappingDf.loc[:, 'ratio_difference'] = (filteredOverlappingDf.count_hs / filteredOverlappingDf.count_expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/2494458605.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CCG</td>\n",
       "      <td>296</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000925</td>\n",
       "      <td>71.089732</td>\n",
       "      <td>4.163752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CATC</td>\n",
       "      <td>263</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>62.203515</td>\n",
       "      <td>4.228057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCAAT</td>\n",
       "      <td>299</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>62.203515</td>\n",
       "      <td>4.806802</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tokens  count_hs  count_ls  proportion_ls  count_expected  ratio_difference\n",
       "8     CCG       296         8       0.000925       71.089732          4.163752\n",
       "10   CATC       263         7       0.000809       62.203515          4.228057\n",
       "6   GCAAT       299         7       0.000809       62.203515          4.806802"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference > 4]\n",
    "overrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rrs</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trnK</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rrl</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mrp</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trnN</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0   rrs     84\n",
       "1  trnK     84\n",
       "2   rrl     78\n",
       "3   mrp     45\n",
       "4  trnN     45"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tRNA-Lys(ttt)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tRNA-Asn(gtt)</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fe-S cluster carrier ATPase%2C Mrp/ApbC/NBP35 ...</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23S ribosomal RNA</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  count\n",
       "0                                      tRNA-Lys(ttt)     84\n",
       "1                                  16S ribosomal RNA     60\n",
       "2                                      tRNA-Asn(gtt)     45\n",
       "3  Fe-S cluster carrier ATPase%2C Mrp/ApbC/NBP35 ...     45\n",
       "4                                  23S ribosomal RNA     43"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented tokens associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1612811/3318705502.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.1]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tokens</th>\n",
       "      <th>count_hs</th>\n",
       "      <th>count_ls</th>\n",
       "      <th>proportion_ls</th>\n",
       "      <th>count_expected</th>\n",
       "      <th>ratio_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>CTCACC</td>\n",
       "      <td>9</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.034924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1089</th>\n",
       "      <td>GACCCC</td>\n",
       "      <td>21</td>\n",
       "      <td>58</td>\n",
       "      <td>0.006707</td>\n",
       "      <td>515.400555</td>\n",
       "      <td>0.040745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401</th>\n",
       "      <td>CCGAG</td>\n",
       "      <td>12</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.046566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1369</th>\n",
       "      <td>TCCCAGC</td>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.050446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1371</th>\n",
       "      <td>CCACCCC</td>\n",
       "      <td>13</td>\n",
       "      <td>26</td>\n",
       "      <td>0.003006</td>\n",
       "      <td>231.041628</td>\n",
       "      <td>0.056267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>AGAGGA</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>0.002775</td>\n",
       "      <td>213.269195</td>\n",
       "      <td>0.056267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>ATGTACA</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>0.001388</td>\n",
       "      <td>106.634598</td>\n",
       "      <td>0.056267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>CCCAAA</td>\n",
       "      <td>19</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.073729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1633</th>\n",
       "      <td>TAGTTTTT</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0.001041</td>\n",
       "      <td>79.975948</td>\n",
       "      <td>0.075023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>CTCTTG</td>\n",
       "      <td>20</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.077610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1597</th>\n",
       "      <td>TATTTAAT</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>0.001156</td>\n",
       "      <td>88.862165</td>\n",
       "      <td>0.078774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>GGAGCAG</td>\n",
       "      <td>24</td>\n",
       "      <td>34</td>\n",
       "      <td>0.003932</td>\n",
       "      <td>302.131360</td>\n",
       "      <td>0.079436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312</th>\n",
       "      <td>CCTAAG</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "      <td>0.002313</td>\n",
       "      <td>177.724329</td>\n",
       "      <td>0.084400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1054</th>\n",
       "      <td>TATCCA</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>0.003353</td>\n",
       "      <td>257.700278</td>\n",
       "      <td>0.085370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>GCTGAG</td>\n",
       "      <td>30</td>\n",
       "      <td>38</td>\n",
       "      <td>0.004394</td>\n",
       "      <td>337.676226</td>\n",
       "      <td>0.088842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>CTTGAT</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>0.001735</td>\n",
       "      <td>133.293247</td>\n",
       "      <td>0.090027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667</th>\n",
       "      <td>GAGGCA</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0.000694</td>\n",
       "      <td>53.317299</td>\n",
       "      <td>0.093778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1483</th>\n",
       "      <td>ATATATTT</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>0.001388</td>\n",
       "      <td>106.634598</td>\n",
       "      <td>0.093778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1619</th>\n",
       "      <td>CTAAGT</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>62.203515</td>\n",
       "      <td>0.096458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1306</th>\n",
       "      <td>TGCAA</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>0.001966</td>\n",
       "      <td>151.065680</td>\n",
       "      <td>0.099295</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tokens  count_hs  count_ls  proportion_ls  count_expected  \\\n",
       "1528    CTCACC         9        29       0.003353      257.700278   \n",
       "1089    GACCCC        21        58       0.006707      515.400555   \n",
       "1401     CCGAG        12        29       0.003353      257.700278   \n",
       "1369   TCCCAGC        13        29       0.003353      257.700278   \n",
       "1371   CCACCCC        13        26       0.003006      231.041628   \n",
       "1399    AGAGGA        12        24       0.002775      213.269195   \n",
       "1607   ATGTACA         6        12       0.001388      106.634598   \n",
       "1161    CCCAAA        19        29       0.003353      257.700278   \n",
       "1633  TAGTTTTT         6         9       0.001041       79.975948   \n",
       "1121    CTCTTG        20        29       0.003353      257.700278   \n",
       "1597  TATTTAAT         7        10       0.001156       88.862165   \n",
       "984    GGAGCAG        24        34       0.003932      302.131360   \n",
       "1312    CCTAAG        15        20       0.002313      177.724329   \n",
       "1054    TATCCA        22        29       0.003353      257.700278   \n",
       "827     GCTGAG        30        38       0.004394      337.676226   \n",
       "1397    CTTGAT        12        15       0.001735      133.293247   \n",
       "1667    GAGGCA         5         6       0.000694       53.317299   \n",
       "1483  ATATATTT        10        12       0.001388      106.634598   \n",
       "1619    CTAAGT         6         7       0.000809       62.203515   \n",
       "1306     TGCAA        15        17       0.001966      151.065680   \n",
       "\n",
       "      ratio_difference  \n",
       "1528          0.034924  \n",
       "1089          0.040745  \n",
       "1401          0.046566  \n",
       "1369          0.050446  \n",
       "1371          0.056267  \n",
       "1399          0.056267  \n",
       "1607          0.056267  \n",
       "1161          0.073729  \n",
       "1633          0.075023  \n",
       "1121          0.077610  \n",
       "1597          0.078774  \n",
       "984           0.079436  \n",
       "1312          0.084400  \n",
       "1054          0.085370  \n",
       "827           0.088842  \n",
       "1397          0.090027  \n",
       "1667          0.093778  \n",
       "1483          0.093778  \n",
       "1619          0.096458  \n",
       "1306          0.099295  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf = filteredOverlappingDf.sort_values(by=['ratio_difference'])[filteredOverlappingDf.ratio_difference < 0.1]\n",
    "underrepresentedTokensDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented genes associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>trnL</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>trnA</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trnR</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>trnP</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>trnG</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene  count\n",
       "0  trnL    113\n",
       "1  trnA     83\n",
       "2  trnR     78\n",
       "3  trnP     76\n",
       "4  trnG     71"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).gene.value_counts().reset_index()[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Under represented annotations associated with the low attribution score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tRNA-Leu(taa)</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tRNA-Ala(tgc)</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tRNA-Pro(tgg)</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tRNA-Arg(acg)</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tRNA-Gly(gcc)</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            name  count\n",
       "0  tRNA-Leu(taa)    113\n",
       "1  tRNA-Ala(tgc)     83\n",
       "2  tRNA-Pro(tgg)     76\n",
       "3  tRNA-Arg(acg)     76\n",
       "4  tRNA-Gly(gcc)     71"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "underrepresentedTokensDf.merge(overlappingDf[['tokens', 'id', 'name', 'gene']], how='inner', on=['tokens']).name.value_counts().reset_index()[:5]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
